---
title: "Ovarioid"
author: "TL"
date: "2023-01-29"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load required library

```{r Load dependencies}
suppressMessages(library(dplyr))
suppressMessages(library(DESeq2))
suppressMessages(library(ggplot2))
suppressMessages(library(pheatmap))
suppressMessages(library(umap))
suppressMessages(library(biomaRt))
suppressMessages(library(org.Hs.eg.db))
suppressMessages(library(DOSE))
suppressMessages(library(pathview))
suppressMessages(library(clusterProfiler))
suppressMessages(library(pheatmap))
suppressMessages(library(VennDiagram))
suppressMessages(library(RColorBrewer))
suppressMessages(library(msigdbr))
suppressMessages(library(edgeR))
suppressMessages(library(stringr))
suppressMessages(library(future))
suppressMessages(library(patchwork))
suppressMessages(library(VennDiagram))
suppressMessages(library(GSEABase))
suppressMessages(library(GSVA))
suppressMessages(library(Seurat))
suppressMessages(library(ComplexHeatmap))

```

## Merge count matrix from individual samples

```{r merge count table}

setwd("/Users/tili/Desktop/PhD/Ovarioid/data/raw_counts/")
path <- "/Users/tili/Desktop/PhD/Ovarioid/data/raw_counts/"

info <- read.table("/Users/tili/Desktop/PhD/Ovarioid/data/info.txt", header = T)

counts <- list.files(path, pattern = "txt", recursive = F)
data <- data.frame(matrix(nrow = 62649, ncol = 1))
for (i in 1:length(counts)) {
  assign(paste0("test", i), read.table(counts[i], header = T))
  data <- cbind(data, get(paste0("test", i)))
  rm(list = ls(pattern = "test"))
}

cells <- data[, -c(1, which(colnames(data) %in% c("Chr", "Start", "End", "Strand", "Length")))]
cells <- cells[, -which(str_detect(colnames(cells), "Geneid."))]
colnames(cells) <- gsub(".Aligned.sortedByCoord.out.bam", "", colnames(cells))

re_name <- function(cells) {
  gene <- cells[,1]
  rownames(cells) <- gene
  cells <- cells[,-1]
  return(cells)
}

cells <- re_name(cells)

d1 <- read.table("/Users/tili/Desktop/PhD/Ovarioid/data/BEAid_sample_name.txt", header = T)
d2 <- read.table("/Users/tili/Desktop/PhD/Ovarioid/data/F.Fagerstrom_22_07_sample_info.txt", header = T)
d2$NGI_ID <- paste0(d2$NGI_ID, "_S", rownames(d2))
d3 <- merge(d1, d2, by.x = "BEA_ID", by.y = "User_ID")
info <- merge(info, d3, by.x = "name", by.y = "NGI_ID")
info <- info %>% mutate(culture = ifelse(str_detect(Customer_ID, "TIS"), "Tissue",
                                      ifelse(str_detect(Customer_ID, "3D"), "3D", "2D"))) %>% 
  mutate(type = ifelse(str_detect(Customer_ID, "COR"), "Cortex", "Medulla")) %>% 
  mutate(freezing = ifelse(str_detect(Customer_ID, "SF"), "SF", "Fresh"))
info$group <- str_sub(info$Customer_ID, end = -3)
info$group <- gsub("_$", "", info$group)

re_order <- function(info, cells) {
  genomic_idx <- match(info$name, colnames(cells))
  genomic_idx
  cells <- cells[ ,genomic_idx]
  all(info$name == colnames(cells))
  return(cells)
}

cells <- re_order(info, cells)

```

## Create dds object

```{r dds object construction}

# correct information with freezing status
info$freezing[c(1,2,5,6,7)] <- rep("Fresh", 5)
info$group[c(1,2,5,6,7)] <- gsub("SF_", "", info$group[c(1,2,5,6,7)])

# remove outliers
info <- info %>% dplyr::filter(name != "P28012_1071_S71")

# remove tissue samples from analysis
info <- info %>% dplyr::filter(culture != "Tissue")

cells <- re_order(info, cells)
info$culture <- factor(info$culture)
info$type <- factor(info$type)
info$freezing <- factor(info$freezing)

dds <- DESeqDataSetFromMatrix(countData=cells, colData = info, design = ~ freezing + type + culture)

# Collapse technical replicates
dds <- collapseReplicates(dds, groupby = dds$group, renameCols = T)

info <- dds@colData %>% as.data.frame()
cells <- dds@assays@data$counts %>% as.data.frame()

rownames(info) <- NULL

write.csv(info, "/Users/tili/Desktop/PhD/Ovarioid/data/processed/info.csv")
write.csv(cells, "/Users/tili/Desktop/PhD/Ovarioid/data/processed/collapsed_counts.csv")

```

## Create dds object

```{r create dds object}

culture <- c("2D", "3D")
freezing <- c("Fresh", "SF")
type <- c("Cortex", "Medulla")

# Remove outlier
info <- info %>% dplyr::filter(name != "P28012_1035_S35")

info$culture <- as.factor(info$culture)
info$freezing <- as.factor(info$freezing)
info$culture <- relevel(info$culture, ref = "2D")

cells <- re_order(info, cells)
print(all(info$name == colnames(cells)))

# Create DESeq2 object
dds <- DESeqDataSetFromMatrix(countData=cells, colData = info, design = ~ freezing + type + culture)


```


## Plot sample counts

```{r filter}

theme <- theme(text=element_text(size=9),
               axis.text.x=element_text(color="black", angle = 90),
               axis.text.y=element_text(color="black"),
               axis.ticks=element_blank(),
               strip.background = element_rect(colour=NA, fill=NA),
               plot.background = element_blank(),
               panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(),
               panel.border = element_blank(),
               axis.line.x = element_line(linewidth=0.4),
               axis.line.y = element_line(linewidth=0.4),
               panel.background = element_blank(),
               legend.position = "bottom")

info %>% mutate(read = colSums(counts(dds))) %>%
            ggplot(aes(x=name, y=read, fill = culture)) + geom_bar(stat = "identity") + theme +
            ggtitle("Samples Sequenced Read") + xlab("") 

```

## Figure 3a, 3b: PCA plot with cpm normalization

```{r PCA cpm}

#info <- read.csv2("/Users/tili/Desktop/PhD/Ovarioid/data/processed/info_Cortex.csv", sep = ",")
#info$culture <- relevel(factor(info$culture), ref = "2D")
r <- edgeR::cpm(counts(dds), normalized.lib.sizes = T, log = T)
pca <- prcomp(t(r))
df <- cbind(info, pca$x)
contri <- round(summary(pca)$importance[2,] * 100, 2)
ggplot(df, aes(label = Customer_ID)) + geom_point(aes(x=PC1, y = PC2, color = culture, shape = type), size = 3) + theme +
      ggtitle("PCA (cpm)") + scale_color_manual(values = c("#6DA7CA","#ECAC27")) +
      xlab(sprintf("PC1 (%s%%)", contri[1])) + ylab(sprintf("PC2 (%s%%)", contri[2])) +
  stat_ellipse(aes(PC1, PC2, group = culture, color = culture))
#plotly::ggplotly(p2)

```

## UMAP plot with cpm normalization

```{r UMAP cpm}

#info <- read.csv2("/Users/tili/Desktop/PhD/Ovarioid/data/processed/info_Cortex.csv", sep = ",")
#info$culture <- relevel(factor(info$culture), ref = "2D")
r <- edgeR::cpm(counts(dds), normalized.lib.sizes = T, log = T)
set.seed(100)
normalized_counts <- t(r)
umap_results <- umap::umap(normalized_counts, n_neighbors = 5) 
umap_plot_df <- data.frame(umap_results$layout) %>%
        tibble::rownames_to_column("name") %>% 
        dplyr::inner_join(info, by = "name")
ggplot(umap_plot_df, aes(x = X1, y = X2, color = culture, shape = type)) + geom_point(size = 3) + theme +
      ggtitle("UMAP (cpm)") + theme + scale_color_manual(values = c("#6DA7CA","#ECAC27")) +
  stat_ellipse(aes(X1, X2, group = culture))

```

## Difference between ovarioid vs 2D

```{r 2D vs 3D}

# Keep genes that have at least 1 read and expressed in at least 5 samples
keep <- rowSums(counts(dds) > 1) >= 5
dds <- dds[keep,]

# Relevel to set the reference as fresh tissue and perform DE analysis
dds$culture <- relevel(dds$culture, ref = "2D")
dds <- DESeq(dds)

# Check the normalization
r <- counts(dds, normalized = T)
boxplot(log10(r))

# Compare 2D and 3D vs Tissue
comparisonALL <- results(dds, contrast = c("culture", "3D", "2D"), 
                              independentFiltering = T, pAdjustMethod = "fdr") %>% 
                              as.data.frame() %>% 
                              mutate(geneid = rownames(.))

#significant DEG
comparison05 <- comparisonALL %>% 
    filter(padj<0.05) %>% 
    filter(abs(log2FoldChange) > 1.5) 

#get gene description
mart <- useEnsembl(biomart = "ensembl",dataset = "hsapiens_gene_ensembl")
output <- getBM(attributes=c("ensembl_gene_id","external_gene_name","entrezgene_id","description"), mart = mart)

get_id <- function(res.sig) {
  res.sig$geneid <- rownames(res.sig)
  final <- merge(res.sig, output, by.x = "geneid", by.y = "ensembl_gene_id", all.x = T, all.y = F)
  return(final)
}

final_all <- get_id(comparisonALL)
final <- get_id(comparison05) %>% unique()

openxlsx::write.xlsx(final, "/Users/tili/Desktop/PhD/Ovarioid/results/tables/Silk-Ov_vs_2D_DESeq2_padj_sig05_FC1.5_20240809.xlsx")

normalized_counts <- counts(dds, normalized=T) %>% 
  data.frame() %>%
  tibble::rownames_to_column(var="gene") %>%
  as_tibble() %>%
  left_join(output, by=c("gene" = "ensembl_gene_id")) %>% 
  unique()

```

## Suppl. Figure S3b: Top500 variable genes with all the samples heatmap

```{r top500 variable genes heatmap}

# Get normalized count matrix
df <- normalized_counts
df <- df[,-c((ncol(df)-2):(ncol(df)))] %>% unique() %>% as.data.frame()
rownames(df) <- df$gene
df <- df[,-1]

# Top500 variable genes
topVarGenes <- head(order(rowVars(as.matrix(df)), decreasing = T), 500)
d1 <- df[topVarGenes,] 
d1 <- na.omit(d1)

# Scale the data
scalecor <- scale(t(d1), scale = T, center = T) %>% t() %>% 
  as.data.frame()

# Get meta data
d2 <- info
d2 <- d2[which(d2$name %in% colnames(scalecor)),]

anno <- d2[, c(1, 6:8)]
rownames(anno) <- anno$name
anno$culture <- factor(anno$culture, levels = c("2D", "3D"))

# Create HeatmapAnnotation object
annot <- HeatmapAnnotation(
  culture = anno$culture,
  type = anno$type,
  #freezing = anno$freezing,
  col = list(
    culture = c("2D" = "#6DA7CA", "3D" = "#ECAC27"),
    type = c("Cortex" = "#EABFBB", "Medulla" = "#91AD5A")#,
    #freezing = c("SF" = "#bf812d", "Fresh" = "#80cdc1")
  )
)

Heatmap(as.matrix(scalecor), name = "Z-score",
        cluster_rows = T,
        show_row_names = F,
        cluster_columns = T,
        clustering_distance_rows = "euclidean",
        column_split = anno$culture,
        top_annotation = annot,
        color = colorRampPalette(c("blue", "white", "red"))(100))

```

## Figure 7b: Cytokine heatmap - 2D removed

```{r cytokine - 2D removed}

cyto <- c("ENSG00000136244", "ENSG00000169429", "ENSG00000108691", "ENSG00000107562", "ENSG00000275302",
          "ENSG00000172156", "ENSG00000277632", "ENSG00000136689", "ENSG00000110944", "ENSG00000164400", 
          "ENSG00000136634")

# Cortex cytokine
df <- normalized_counts
id_c <- match(cyto, df$gene) %>% na.omit()
d1 <- df[id_c,] 
d1 <- na.omit(d1) %>% as.data.frame()
rownames(d1) <- d1$external_gene_name
d1 <- d1[!duplicated(d1$external_gene_name),]
d2 <- info %>% filter(culture != "2D")
genes_c <- d1[,(ncol(d1)-2)]
d1 <- d1[,which(colnames(d1) %in% d2$name)]

scalecor <- scale(t(d1), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_c) %>% 
  na.omit()
rownames(scalecor) <- scalecor$genes_c

d2 <- d2[which(d2$name %in% colnames(scalecor)),]

anno <- d2[, c(1, 6:8)]
rownames(anno) <- anno$name

# Create HeatmapAnnotation object
annot <- HeatmapAnnotation(
  #culture = anno$culture,
  type = anno$type,
  #freezing = anno$freezing,
  col = list(
    #culture = c("2D" = "#6DA7CA", "3D" = "#ECAC27"),
    type = c("Cortex" = "#EABFBB", "Medulla" = "#91AD5A")#,
    #freezing = c("SF" = "#bf812d", "Fresh" = "#80cdc1")
  )
)

color_scale <- colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))
Heatmap(as.matrix(scalecor[,-ncol(scalecor)]), name = "Z-score",
        color = color_scale,
        cluster_rows = T,
        show_row_names = T,
        cluster_columns = T,
        column_split = anno$type,
        clustering_distance_rows = "euclidean",
        top_annotation = annot)

############################## Barplot with mean + sem

# Barplot of all genes, d1 is the normalized counts
d3 <- d1 %>% mutate(cytokine = rownames(.)) %>% 
  pivot_longer(!cytokine, names_to = "sample", values_to = "value")

# Check if the samples are the same in the metadata
all(d3$sample[1:10] == anno$name)

# Add meta data
d3$type <- rep(anno$type, 11)

# Cortex
cort <- d3 %>% dplyr::filter(type == "Cortex") %>% arrange(desc(value))
orders <- unique(cort$cytokine)
cort$cytokine <- factor(cort$cytokine, levels = orders)

cort %>% filter(type == "Cortex") %>% 
  group_by(cytokine) %>% mutate(average = mean(value)) %>% mutate(sem = sd(value)/sqrt(5)) %>% 
  ggplot(aes(x = as.numeric(cytokine), y = average)) + 
  geom_col(position = "dodge", width = 0.4, fill = "#EABFBB") + theme_bw() +
  geom_errorbar(aes(ymin = average - sem, ymax = average + sem), position = position_dodge(0.1), width = 0.2, color = "black") +
  ggtitle("Cytokines - Cortex") + #ylim(c(0, 5000)) +
  facet_zoom(xy = cort$cytokine %in% c("IL23A", "IL1RN",    
                                         "CSF2", "CCL4", 
                                         "CCL3"), horizontal=FALSE) +
  scale_x_continuous(breaks = 1:length(levels(cort$cytokine)),
                     label = levels(cort$cytokine)) + 
  theme(legend.position = c(0.90, 0.90), legend.background = element_rect(fill="transparent"),
        legend.title = element_blank(), axis.text.x=element_text(angle=55, vjust=1,  hjust=1,size = 8)) +
  xlab("Cytokines") + ylab("Normalized Counts")

# Medulla
med <- d3 %>% dplyr::filter(type == "Medulla") %>% arrange(desc(value))
orders <- unique(med$cytokine)
med$cytokine <- factor(med$cytokine, levels = orders)

med %>% filter(type == "Medulla") %>% 
  group_by(cytokine) %>% mutate(average = mean(value)) %>% mutate(sem = sd(value)/sqrt(5)) %>% 
  ggplot(aes(x = as.numeric(cytokine), y = average)) + 
  geom_col(position = "dodge", width = 0.4, fill = "#91AD5A") + theme_bw() +
  geom_errorbar(aes(ymin = average - sem, ymax = average + sem), position = position_dodge(0.1), width = 0.2, color = "black") +
  ggtitle("Cytokines - Medulla") + #ylim(c(0, 5000)) +
  facet_zoom(xy = med$cytokine %in% c("IL10", "IL23A", "IL1RN",    
                                         "CSF2", "CCL4", 
                                         "CCL3"), horizontal=FALSE) +
  scale_x_continuous(breaks = 1:length(levels(med$cytokine)),
                     label = levels(med$cytokine)) + 
  theme(legend.position = c(0.90, 0.90), legend.background = element_rect(fill="transparent"),
        legend.title = element_blank(), axis.text.x=element_text(angle=55, vjust=1,  hjust=1,size = 8)) +
  xlab("Cytokines") + ylab("Normalized Counts")

```

## Figure 3c: Cellcycle genes heatmap

```{r cell cycle genes}

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

cycle <- c(s.genes, g2m.genes)

# Get normalized counts for cell-cycle genes
df <- normalized_counts
id_c <- match(cycle, df$external_gene_name)
d1 <- df[id_c,] 
d1 <- na.omit(d1)

genes_c <- d1[,(ncol(d1)-2)]

scalecor <- scale(t(d1[,2:(ncol(d1)-3)]), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_c) %>% 
  na.omit()
rownames(scalecor) <- scalecor$external_gene_name

d2 <- info
d2 <- d2[which(d2$name %in% colnames(scalecor)),]

anno <- d2[, c(1, 6:8)]
rownames(anno) <- anno$name

d5 <- anno %>% filter(culture == "2D") # %>% filter(type == "Cortex")
c2D <- scalecor[,which(colnames(scalecor) %in% d5$name)]
c2DD <- apply(c2D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(c2D)) %>% 
  'colnames<-' ("2D")

d6 <- anno %>% filter(culture == "3D") #%>% filter(type == "Cortex")
c3D <- scalecor[,which(colnames(scalecor) %in% d6$name)]
c3DD <- apply(c3D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(c3D)) %>% 
  'colnames<-' ("3D")

con <- cbind(c2DD, c3DD)

# Get the data frame so that the heatmap can be splited into S phase and G2M phase genes
cellcycle <- cycle %>% as.data.frame() %>% 'colnames<-' ("Gene")
cellcycle$group <- ifelse(cellcycle$Gene %in% s.genes, "S-phase", "G2M-phase")
cellcycle <- cellcycle[which(cellcycle$Gene %in% rownames(c2D)),] 

Heatmap(as.matrix(con), name = "Z-score",
        cluster_row_slices = T,
        cluster_columns = F,
        show_row_names = F,
        clustering_distance_rows = "euclidean",
        row_split = cellcycle$group,
        row_gap = unit(2, "mm"),
        color = colorRampPalette(c("blue", "white", "red"))(100))

```

## Figure 3e: Angiogenesis genes heatmap

```{r Angiogenesis genes}

angiogenesis <- bitr("GO:0001525", fromType = "GO", toType = "SYMBOL", OrgDb = org.Hs.eg.db)
angio <- angiogenesis$SYMBOL

df <- normalized_counts
id_c <- match(angio, df$external_gene_name) %>% na.omit()
d1 <- df[id_c,] 
d1 <- na.omit(d1)
d1 <- d1[!duplicated(d1$external_gene_name),]

genes_c <- d1[,(ncol(d1)-2)]

scalecor <- scale(t(d1[,2:(ncol(d1)-3)]), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_c) %>% 
  na.omit()
rownames(scalecor) <- scalecor$external_gene_name

d2 <- info
d2 <- d2[which(d2$name %in% colnames(scalecor)),]

anno <- d2[, c(1, 6:8)]
rownames(anno) <- anno$name

d5 <- anno %>% filter(culture == "2D") # %>% filter(type == "Cortex")
c2D <- scalecor[,which(colnames(scalecor) %in% d5$name)]
c2DD <- apply(c2D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(c2D)) %>% 
  'colnames<-' ("2D")

d6 <- anno %>% filter(culture == "3D") #%>% filter(type == "Cortex")
c3D <- scalecor[,which(colnames(scalecor) %in% d6$name)]
c3DD <- apply(c3D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(c3D)) %>% 
  'colnames<-' ("3D")

# Combine 2D and 3D but keep the z score
con <- cbind(c2DD, c3DD)

color_scale <- colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))
heatmap <- Heatmap(as.matrix(con), name = "Z-score",
        column_title = "Angiogenesis",
        cluster_columns = F,
        clustering_distance_rows = "euclidean",
        row_km = 2,
        #row_gap = unit(2, "mm"),
        show_row_names = F,
        color = color_scale)

# Extract row order for each cluster
row_order_list <- row_order(heatmap)

# Retrieve row names for each cluster
clustered_row_names <- lapply(row_order_list, function(cluster) rownames(con)[cluster])

# Perform GO enrichment analysis
cluster1 <- bitr(clustered_row_names[[1]], fromType = "SYMBOL", toType = "ENTREZID",
                 OrgDb = org.Hs.eg.db)

ego1 <- enrichGO(gene = cluster1$ENTREZID,
                OrgDb = org.Hs.eg.db,
                keyType = "ENTREZID",
                ont = "BP",  
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)
d7 <- ego1@result %>% mutate(cluster = rep("Cluster1", nrow(.))) %>% 
  arrange(p.adjust) %>% head(n = 20)

# Cluster 2
cluster2 <- bitr(clustered_row_names[[2]], fromType = "SYMBOL", toType = "ENTREZID",
                 OrgDb = org.Hs.eg.db)

ego2 <- enrichGO(gene = cluster2$ENTREZID,
                OrgDb = org.Hs.eg.db,
                keyType = "ENTREZID",
                ont = "BP",  
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)
d8 <- ego2@result %>% mutate(cluster = rep("Cluster2", nrow(.))) %>% 
  arrange(p.adjust) %>% head(n = 20)

annotation_data <- rbind(d7, d8)

# Integrate GO results to heatmap
category_colors <- c("Cluster1" = "steelblue", "Cluster2" = "orange")
draw(rowAnnotation(barplot = anno_barplot(-log10(annotation_data$p.adjust), ylim = c(0, 40),
    gp = gpar(fill = category_colors[annotation_data$cluster], alpha = 0.5),
    width = unit(8, "cm")  # Set the width of the barplot
  ), show_annotation_name = TRUE) + 
  rowAnnotation(GO = anno_text(annotation_data$Description, 
                               gp = gpar(col = "black", fontsize = 9),
                               location = -1.15,
                               just = "left")))


```

## Figure 3d: Hypoxia genes heatmap

```{r Hypoxia genes}

read_excel_allsheets <- function(filename, tibble = FALSE) {
    sheets <- readxl::excel_sheets(filename)
    x <- lapply(sheets, function(X) readxl::read_excel(filename, sheet = X))
    if(!tibble) x <- lapply(x, as.data.frame)
    names(x) <- sheets
    x
}

kgn <- read_excel_allsheets("/Users/tili/Desktop/PhD/Ovarioid/results/hypoxia_genes.xlsx")
hypoxia <- kgn[[1]]

# Cortex hypoxia
df <- normalized_counts
id_c <- match(hypoxia$gene, df$external_gene_name) %>% na.omit()
d1 <- df[id_c,] 
d1 <- na.omit(d1)

genes_c <- d1[,(ncol(d1)-2)]

scalecor <- scale(t(d1[,2:(ncol(d1)-3)]), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_c) %>% 
  na.omit()
rownames(scalecor) <- scalecor$external_gene_name

d2 <- info
d2 <- d2[which(d2$name %in% colnames(scalecor)),]

anno <- d2[, c(1, 6:8)]
rownames(anno) <- anno$name

d5 <- anno %>% filter(culture == "2D") # %>% filter(type == "Cortex")
c2D <- scalecor[,which(colnames(scalecor) %in% d5$name)]
c2DD <- apply(c2D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(c2D)) %>% 
  'colnames<-' ("2D")

d6 <- anno %>% filter(culture == "3D") #%>% filter(type == "Cortex")
c3D <- scalecor[,which(colnames(scalecor) %in% d6$name)]
c3DD <- apply(c3D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(c3D)) %>% 
  'colnames<-' ("3D")

# Combine cortex and medulla but keep the z score
con <- cbind(c2DD, c3DD)

heatmap <- Heatmap(as.matrix(con), name = "Z-score",
        column_title = "Hypoxia",
        cluster_columns = F,
        clustering_distance_rows = "euclidean",
        row_km = 2,
        #row_gap = unit(2, "mm"),
        show_row_names = F,
        column_split = anno_group$group,
        color = colorRampPalette(c("blue", "white", "red"))(100))

# Extract row order for each cluster
row_order_list <- row_order(heatmap)

# Retrieve row names for each cluster
clustered_row_names <- lapply(row_order_list, function(cluster) rownames(con)[cluster])

# Perform GO enrichment analysis
cluster1 <- bitr(clustered_row_names[[1]], fromType = "SYMBOL", toType = "ENTREZID",
                 OrgDb = org.Hs.eg.db)

ego1 <- enrichGO(gene = cluster1$ENTREZID,
                OrgDb = org.Hs.eg.db,
                keyType = "ENTREZID",
                ont = "BP",  
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)
d7 <- ego1@result %>% mutate(cluster = rep("Cluster1", nrow(.))) %>% 
  arrange(p.adjust) %>% head(n = 30)

# Cluster 2
cluster2 <- bitr(clustered_row_names[[2]], fromType = "SYMBOL", toType = "ENTREZID",
                 OrgDb = org.Hs.eg.db)

ego2 <- enrichGO(gene = cluster2$ENTREZID,
                OrgDb = org.Hs.eg.db,
                keyType = "ENTREZID",
                ont = "BP",  
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)
d8 <- ego2@result %>% mutate(cluster = rep("Cluster2", nrow(.))) %>% 
  arrange(p.adjust) %>% head(n = 30)

annotation_data <- rbind(d7, d8)

# Integrate GO results to heatmap
category_colors <- c("Cluster1" = "steelblue", "Cluster2" = "orange")
draw(rowAnnotation(barplot = anno_barplot(-log10(annotation_data$p.adjust), ylim = c(0, 250),
    gp = gpar(fill = category_colors[annotation_data$cluster], alpha = 0.5),
    width = unit(10, "cm")  # Set the width of the barplot
  ), show_annotation_name = TRUE) + 
  rowAnnotation(GO = anno_text(annotation_data$Description, 
                               gp = gpar(col = "black", fontsize = 9),
                               location = -1.1,
                               just = "left")))

```

## Suppl. Figure S4: Steroidogenesis genes heatmap - 2D removed

```{r steroidogenesis genes - 2D removed}

kgn <- read.csv("/Users/tili/Desktop/PhD/Ovarioid/data/steroids.csv", sep = ",")
steroid <- kgn$geneID

# Cortex steroidogenesis
df <- normalized_counts[["Cortex"]]
id_c <- match(steroid, df$gene) %>% na.omit()
d1 <- df[id_c,] 
d1 <- na.omit(d1) %>% as.data.frame()
rownames(d1) <- d1$external_gene_name
d2 <- cor_df %>% filter(culture != "2D")
genes_c <- d1[,(ncol(d1)-2)]
d1 <- d1[,which(colnames(d1) %in% d2$name)]

scalecor <- scale(t(d1), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_c) %>% 
  na.omit()
rownames(scalecor) <- scalecor$genes_c

d2 <- d2[which(d2$name %in% colnames(scalecor)),]

anno <- d2[, c(1, 2, 7, 11)]
rownames(anno) <- anno$name

pheatmap(scalecor[, -ncol(scalecor)], cluster_rows = F, show_rownames = T, 
         cluster_cols = T, clustering_distance_rows = "correlation", annotation = anno, 
         color = colorRampPalette(c("blue", "white", "red"))(100),
         breaks = seq(-1, 1, length.out = 90))

d4 <- anno %>% filter(culture == "Tissue") #%>% filter(type == "Cortex")
tis <- scalecor[,which(colnames(scalecor) %in% d4$name)]
tiss <- apply(tis, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(tis)) %>% 
  'colnames<-' ("Tissue")

d6 <- anno %>% filter(culture == "3D") #%>% filter(type == "Cortex")
c3D <- scalecor[,which(colnames(scalecor) %in% d6$name)]
c3DD <- apply(c3D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(c3D)) %>% 
  'colnames<-' ("3D")


# Medulla steroidogenesis
df <- normalized_counts[["Medulla"]]
id_m <- match(steroid, df$gene) %>% na.omit()
d1 <- df[id_m,] 
d1 <- na.omit(d1) %>% as.data.frame()
rownames(d1) <- d1$external_gene_name
d2 <- med_df %>% filter(culture != "2D")
genes_m <- d1[,(ncol(d1)-2)]
d1 <- d1[,which(colnames(d1) %in% d2$name)]

scalemed <- scale(t(d1), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_m) %>% 
  na.omit()
rownames(scalemed) <- scalemed$genes_m

d2 <- d2[which(d2$name %in% colnames(scalemed)),]

anno <- d2[, c(1, 2, 7, 11)]
rownames(anno) <- anno$name

pheatmap(scalemed[, -ncol(scalemed)], cluster_rows = F, show_rownames = T, 
         cluster_cols = T, clustering_distance_rows = "correlation", annotation = anno, 
         color = colorRampPalette(c("blue", "white", "red"))(100),
         breaks = seq(-1, 1, length.out = 100))

d7 <- anno %>% filter(culture == "Tissue") #%>% filter(type == "Medulla")
mtis <- scalemed[,which(colnames(scalemed) %in% d7$name)]
mtiss <- apply(mtis, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(mtis)) %>% 
  'colnames<-' ("Tissue_med")

d9 <- anno %>% filter(culture == "3D") #%>% filter(type == "Medulla")
m3D <- scalemed[,which(colnames(scalemed) %in% d9$name)]
m3DD <- apply(m3D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(m3D)) %>% 
  'colnames<-' ("3D_med")

# Combine cortex and medulla but keep the z score
common <- intersect(rownames(tiss), rownames(mtiss))

cortex <- cbind(tiss, c3DD)
cortex <- cortex[which(rownames(cortex) %in% common),]

medulla <- cbind(mtiss, m3DD)
medulla <- medulla[which(rownames(medulla) %in% common),]
con <- cbind(cortex, medulla)

anno_group <- colnames(con) %>% as.data.frame()
anno_group$group <- rep(c("Cortex", "Medulla"), each = 2)

Heatmap(as.matrix(con), name = "Z-score",
        #cluster_row_slices = T,
        cluster_columns = F,
        clustering_distance_rows = "euclidean",
        row_km = 2,
        row_gap = unit(1, "mm"),
        show_row_names = T,
        column_split = anno_group$group,
        color = colorRampPalette(c("blue", "white", "red"))(100))

```


## Difference between different culture condition

```{r shrink for functional analysis}

res_shrunk <- lfcShrink(dds, coef =  match(paste0("culture_3D_vs_2D"),   
                             resultsNames(dds)), 
                             type = "apeglm") %>% as.data.frame() %>% 
                             mutate(geneid = rownames(.))

shrunk_sig <- res_shrunk %>% 
      filter(padj < 0.05)  

res_shrunk <- get_id(res_shrunk)
shrunk_sig <- get_id(shrunk_sig)

```

## Gene set analysis
## Figure 7a, 7b: Gene-concept network

```{r hallmark gene set analysis}

msigHs_h <- msigdbr(species = "Homo sapiens", category = "H") %>%
  select(gs_name, entrez_gene) %>% 
  as.data.frame()

msigHs_c2 <- msigdbr(species = "Homo sapiens", category = "C2") %>%
  select(gs_name, entrez_gene) %>% 
  as.data.frame()

genelist <- shrunk_sig %>% dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    mutate(logpadj = ifelse(log2FoldChange<0, log10(padj), -log10(padj))) %>% 
    filter(!is.na(entrezgene_id)) %>% 
    unique()
GeneList <- genelist$log2FoldChange
names(GeneList) <- as.character(genelist$entrezgene_id)
GeneList <- sort(GeneList, decreasing = T) 

# Remove duplicated genes in the list
GeneList <- GeneList[-which(duplicated(GeneList))]

edo <- GSEA(geneList = GeneList, minGSSize = 10, maxGSSize = 500,
            pvalueCutoff = 1,
            pAdjustMethod = "BH",
            TERM2GENE = msigHs_h)

edo <- setReadable(edo, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
d9 <- edo@result %>% arrange(desc(NES))

openxlsx::write.xlsx(GSEA_hallmark, file = "/Users/tianyili/Desktop/Ovarioid/results/tables/DESeq2_S71_removed_gsea_hallmark_20230731.xlsx")

selected <- c("HALLMARK_HYPOXIA", "HALLMARK_ANGIOGENESIS")
cnetplot(GSEA_hallmark[["Cortex_3D_vs_Tissue"]], showCategory = selected, 
         foldChange = GeneList[["Cortex_3D_vs_Tissue"]]) +
  ggtitle("Cortex")
cnetplot(GSEA_hallmark[["Medulla_3D_vs_Tissue"]], showCategory = selected, foldChange = GeneList[["Medulla_3D_vs_Tissue"]]) +
  ggtitle("Medulla")

GSEA_c2 <- list()
GSEA_c2_df <- list()
for(i in 1:length(GeneList)){
  print(names(GeneList)[i])
  edo <- GSEA(geneList = GeneList[[i]], minGSSize = 10, maxGSSize = 500,
            pvalueCutoff = 1,
            pAdjustMethod = "BH",
            TERM2GENE = msigHs_c2)
  GSEA_c2[[names(GeneList)[i]]] <- edo 
  GSEA_c2_df[[names(GeneList)[i]]] <- edo %>% as.data.frame()
}

for (i in 1:length(GSEA_c2)) {
  GSEA_c2[[i]] <- setReadable(GSEA_c2[[i]], OrgDb = org.Hs.eg.db, keyType="ENTREZID")
}

openxlsx::write.xlsx(GSEA_c2, file = "/Users/tianyili/Desktop/Ovarioid/results/tables/DESeq2_S71_removed_gsea_c2_20230731.xlsx")

```

## Figure 3a, 3b: Plot GSEA results

```{r GSEA plotting}

GSEA_hallmark_df <- read_excel_allsheets("/Users/tianyili/Desktop/Ovarioid/results/tables/DESeq2_S71_removed_gsea_hallmark_20231130.xlsx")

GSEA_hallmark_df <- lapply(GSEA_hallmark_df, function(x) {
  x <- x %>% dplyr::filter(p.adjust < 0.05)
})

deg_df <- data.frame()
for (i in names(GSEA_hallmark_df)){
  deg <- GSEA_hallmark_df[[i]] %>% mutate(group = rep(i, nrow(GSEA_hallmark_df[[i]])))
  deg_df <- rbind(deg_df, deg) 
}

set <- deg_df %>% 
  mutate(type = (ifelse(str_detect(group, "Cortex"), "Cortex", "Medulla"))) %>% 
  mutate(culture = (ifelse(str_detect(group, "2D"), "2D", "3D")))

# remove 3D vs 2D medulla
set <- set[-c(81:82),]

d5 <- set %>% select(ID, NES, group, type, culture)

d6 <- d5 %>% filter(type == "Cortex")
d6 <- d5 %>% filter(type == "Medulla")
d6$regulate <- ifelse(str_detect(d6$NES, "-"), "Down", "Up")
d6$score <- abs(d6$NES)
d6$score <- ifelse(str_detect(d6$culture, "2D"), as.numeric(paste0("-", d6$score)), d6$score)
all <- d6$ID %>% unique()

# Cortex dataset
temp3 <- d6[1:19,] %>% arrange(desc(score)) # take the unique ID for 2D dataset and arrange by score
orders <- temp3$ID
setdiff(all, orders)
# Add the gene set not in the previous sorting
orders <- c(orders, "HALLMARK_KRAS_SIGNALING_UP", "HALLMARK_ALLOGRAFT_REJECTION",
            "HALLMARK_APOPTOSIS", "HALLMARK_ANDROGEN_RESPONSE", "HALLMARK_IL6_JAK_STAT3_SIGNALING",
            "HALLMARK_IL2_STAT5_SIGNALING")
d6$ID <- factor(d6$ID, levels = orders)

# Medulla dataset
temp4 <- d6[1:21,] %>% arrange(desc(score))
orders <- temp4$ID
setdiff(all, orders)
orders <- c(orders, "HALLMARK_KRAS_SIGNALING_UP", 
            "HALLMARK_INTERFERON_GAMMA_RESPONSE")
d6$ID <- factor(d6$ID, levels = orders)

ggplot(d6, aes(x = ID, y = score, fill = regulate)) + 
  geom_bar(position = "dodge", stat = "identity") +
  scale_fill_manual(values = c( "#ADD187", "#8481BA")) +
  geom_hline(yintercept = 0, color = "black", size = 1) +
  coord_flip() + ylab("NES") + xlab("") +
  ggtitle("GSEA analysis - Medulla") +
  theme(text=element_text(size=15),
               axis.text.x=element_text(color="black", angle = 90, vjust = 0.3, hjust = 0.3),
               axis.text.y=element_text(color="black"),
               axis.ticks.x.bottom=element_blank(),
               strip.background = element_rect(colour=NA, fill=NA),
               plot.background = element_blank(),
               panel.spacing.x=unit(.5, "lines"),
               panel.border = element_rect(color = "black", fill = NA, linewidth = 1),
               axis.line.x = element_line(linewidth=0.4),
               axis.line.y = element_line(linewidth=0.4),
               legend.key.size = unit(2, 'lines'),
               panel.background = element_blank()
)

```

## Venn diagram

```{r venn diagram}

VennDiagram::venn.diagram(
  x = list(na.omit(comparison05[[1]]$geneid), na.omit(comparison05[[2]]$geneid)),
  category.names = c("2D" , "3D"),
  filename = 'Venn for cortex 2D and 3D vs Tissue overlap sig05 FC2.png',
  #filename = NULL,
  output=TRUE,
  # Circles
        lwd = 2,
        col=c("#8dd3c7","#fb8072"),
        fill = c(alpha("#8dd3c7",1),alpha("#fb8072",1)),
        scaled = F,
        cex = 1.5,
        fontface = "bold",
        fontfamily = "sans",
        # Set names
        cat.cex = 0.8,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans",
        cat.col =  c("#8dd3c7","#fb8072")
)


```

## Venn diagram of GO hypoxia, cell cycle, angiogenesis and apoptotic signaling pathway

```{r Venn diagram of GO terms}

apoptosis <- bitr("GO:0097190", fromType = "GO", toType = "SYMBOL", OrgDb = org.Hs.eg.db)

kgn <- read_excel_allsheets("/Users/tianyili/Desktop/Ovarioid/results/hypoxia_genes.xlsx")
hypoxia <- kgn[[1]]

angiogenesis <- bitr("GO:0001525", fromType = "GO", toType = "SYMBOL", OrgDb = org.Hs.eg.db)
angio <- angiogenesis$SYMBOL

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
cycle <- c(s.genes, g2m.genes)

VennDiagram::venn.diagram(
  x = list(apoptosis$SYMBOL, cycle, hypoxia$gene, angio),
  category.names = c("Apoptosis" , "Cell Cycle", "Hypoxia", "Angiogenesis"),
  filename = 'Venn for GO term overlap.png',
  #filename = NULL,
  output=TRUE,
  # Circles
        lwd = 2,
        col=c("#8dd3c7","#fdb462", "#bebada", "#80b1d3"),
        fill = c(alpha("#8dd3c7",1),alpha("#fdb462",1), alpha("#bebada",1), alpha("#80b1d3",1)),
        scaled = F,
        cex = 1.5,
        fontface = "bold",
        fontfamily = "sans",
        # Set names
        cat.cex = 0.8,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans",
        cat.col =  c("#8dd3c7","#fdb462", "#bebada", "#80b1d3")
)

# Hallmark gene set overlapping

angiog <- read.delim("/Users/tianyili/Downloads/HALLMARK_ANGIOGENESIS.v2023.2.Hs.tsv")
angio <- str_split(angiog[which(angiog$STANDARD_NAME == "GENE_SYMBOLS"), 2], ",", simplify = F) %>% unlist()

apoptosis <- read.delim("/Users/tianyili/Downloads/HALLMARK_APOPTOSIS.v2023.2.Hs.tsv")
apop <- str_split(apoptosis[which(apoptosis$STANDARD_NAME == "GENE_SYMBOLS"), 2], ",", simplify = F) %>% unlist()

g2m <- read.delim("/Users/tianyili/Downloads/HALLMARK_G2M_CHECKPOINT.v2023.2.Hs.tsv")
g2m <- str_split(g2m[which(g2m$STANDARD_NAME == "GENE_SYMBOLS"), 2], ",", simplify = F) %>% unlist()

hypoxia <- read.delim("/Users/tianyili/Downloads/HALLMARK_HYPOXIA.v2023.2.Hs.tsv")
hypo <- str_split(hypoxia[which(hypoxia$STANDARD_NAME == "GENE_SYMBOLS"), 2], ",", simplify = F) %>% unlist()

VennDiagram::venn.diagram(
  x = list(apop, g2m, hypo, angio),
  category.names = c("Apoptosis" , "G2M checkpoint", "Hypoxia", "Angiogenesis"),
  filename = 'Venn for Hallmark overlap.png',
  #filename = NULL,
  output=TRUE,
  # Circles
        lwd = 2,
        col=c("#8dd3c7","#fdb462", "#bebada", "#80b1d3"),
        fill = c(alpha("#8dd3c7",1),alpha("#fdb462",1), alpha("#bebada",1), alpha("#80b1d3",1)),
        scaled = F,
        cex = 1.5,
        fontface = "bold",
        fontfamily = "sans",
        # Set names
        cat.cex = 0.8,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans",
        cat.col =  c("#8dd3c7","#fdb462", "#bebada", "#80b1d3")
)

```


## GSEA for 3D specific DEGs

```{r gsea 3D specific DEGs}

common <- intersect(final[[1]]$geneid, final[[2]]$geneid)
deg_3d <- final[[2]][-which(final[[2]]$geneid %in% common),] 

genelist <- deg_3d %>% dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    filter(!is.na(entrezgene_id))
gl <- genelist$log2FoldChange
names(gl) <- as.character(genelist$entrezgene_id)
gl <- sort(gl, decreasing = T)

edo <- GSEA(geneList = gl, minGSSize = 10, maxGSSize = 500,
            pvalueCutoff = 1,
            pAdjustMethod = "BH",
            TERM2GENE = msigHs_h)

edo <- setReadable(edo, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
result <- edo@result %>% as.data.frame()

# 2D specific DEGs enrichment
deg_2d <- final[[1]][-which(final[[1]]$geneid %in% common),] 

genelist <- deg_2d %>% dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    filter(!is.na(entrezgene_id))
gl <- genelist$log2FoldChange
names(gl) <- as.character(genelist$entrezgene_id)
gl <- sort(gl, decreasing = T)

edo <- GSEA(geneList = gl, minGSSize = 10, maxGSSize = 500,
            pvalueCutoff = 1,
            pAdjustMethod = "BH",
            TERM2GENE = msigHs_h)

edo <- setReadable(edo, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
result2 <- edo@result %>% as.data.frame()

write.csv(result, file = "/Users/tianyili/Desktop/Ovarioid/results/tables/GSEA_3D_specific_DEGs_cortex_20230731.csv")
write.csv(result2, file = "/Users/tianyili/Desktop/Ovarioid/results/tables/GSEA_2D_specific_DEGs_cortex_20230731.csv")

#### Medulla

common <- intersect(final[[3]]$geneid, final[[4]]$geneid)
deg_3d <- final[[4]][-which(final[[4]]$geneid %in% common),] 

genelist <- deg_3d %>% dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    filter(!is.na(entrezgene_id))
gl <- genelist$log2FoldChange
names(gl) <- as.character(genelist$entrezgene_id)
gl <- sort(gl, decreasing = T)

edo <- GSEA(geneList = gl, minGSSize = 10, maxGSSize = 500,
            pvalueCutoff = 1,
            pAdjustMethod = "BH",
            TERM2GENE = msigHs_h)

edo <- setReadable(edo, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
result <- edo@result %>% as.data.frame()

# 2D specific DEGs enrichment
deg_2d <- final[[3]][-which(final[[3]]$geneid %in% common),] 

genelist <- deg_2d %>% dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    filter(!is.na(entrezgene_id))
gl <- genelist$log2FoldChange
names(gl) <- as.character(genelist$entrezgene_id)
gl <- sort(gl, decreasing = T)

edo <- GSEA(geneList = gl, minGSSize = 10, maxGSSize = 500,
            pvalueCutoff = 1,
            pAdjustMethod = "BH",
            TERM2GENE = msigHs_h)

edo <- setReadable(edo, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
result2 <- edo@result %>% as.data.frame()

write.csv(result, file = "/Users/tianyili/Desktop/Ovarioid/results/tables/GSEA_3D_specific_DEGs_medulla_20230731.csv")
write.csv(result2, file = "/Users/tianyili/Desktop/Ovarioid/results/tables/GSEA_2D_specific_DEGs_medulla_20230731.csv")


```

## enrichKEGG and gseKEGG

```{r enrich KEGG}

KEGG_enrich <- list()
for(i in 1:length(shrunk_sig)){
  print(names(shrunk_sig)[i])
  KEGG_enrich[[names(shrunk_sig)[i]]] <- enrichKEGG(gene = shrunk_sig[[i]]$entrezgene_id, 
                       organism = 'hsa',
                       pvalueCutoff = 0.1)
}

for (i in 1:length(KEGG_enrich)) {
  KEGG_enrich[[i]] <- setReadable(KEGG_enrich[[i]], OrgDb = org.Hs.eg.db, keyType="ENTREZID")
}

openxlsx::write.xlsx(KEGG_enrich, "/Users/tianyi.li.2/Desktop/Ovarioid/results/tables/DESeq2_S71_removed_sigDEGs_enricher_KEGG_20230222.xlsx")

KEGG_gsea <- list()
for(i in 1:length(GeneList)){
  print(names(GeneList)[i])
  KEGG_gsea[[names(GeneList)[i]]] <- gseKEGG(geneList = GeneList[[i]], 
               organism     = 'hsa',
               minGSSize    = 10,
               pvalueCutoff = 0.1,
               verbose      = FALSE)
}

for (i in 1:length(KEGG_gsea)) {
  KEGG_gsea[[i]] <- setReadable(KEGG_gsea[[i]], OrgDb = org.Hs.eg.db, keyType="ENTREZID")
}

openxlsx::write.xlsx(KEGG_gsea, "/Users/tianyi.li.2/Desktop/Ovarioid/results/tables/DESeq2_S71_removed_sigDEGs_GSEA_KEGG_20230222.xlsx")

```

## GO analysis

```{r GO analysis}

GO_enrich <- list()
for(i in 1:length(shrunk_sig)){
  print(names(shrunk_sig)[i])
  GO_enrich[[names(shrunk_sig)[i]]] <- enrichGO(gene = shrunk_sig[[i]]$entrezgene_id, 
                                                     universe = res_shrunk[[i]]$entrezgene_id,
                                                     OrgDb = org.Hs.eg.db, ont = "ALL", 
                                                     pAdjustMethod = "BH", 
                                                     pvalueCutoff = 0.05, readable = TRUE)
}

for (i in 1:length(GO_enrich)) {
  GO_enrich[[i]] <- setReadable(GO_enrich[[i]], OrgDb = org.Hs.eg.db, keyType="ENTREZID")
}

openxlsx::write.xlsx(GO_enrich, "/Users/tianyili/Desktop/Ovarioid/results/tables/DESeq2_S71_removed_sigDEGs_enrichGO_20230731.xlsx")

gse_GO <- list()
gse_GO_df <- list()
for(i in 1:length(GeneList)){
  print(names(GeneList)[i])
  edo <- gseGO(geneList = GeneList[[i]], 
              OrgDb        = org.Hs.eg.db,
              ont          = "BP",
              minGSSize    = 10,
              maxGSSize    = 500,
              pvalueCutoff = 1,
              verbose      = FALSE)
  gse_GO[[names(GeneList)[i]]] <- edo 
  gse_GO_df[[names(GeneList)[i]]] <- edo %>% as.data.frame()
}

for (i in 1:length(gse_GO)) {
  gse_GO[[i]] <- setReadable(gse_GO[[i]], OrgDb = org.Hs.eg.db, keyType="ENTREZID")
}

openxlsx::write.xlsx(gse_GO, "/Users/tianyili/Desktop/Ovarioid/results/tables/DESeq2_S71_removed_sigDEGs_gseGO_20231217.xlsx")

```

## Plot GO analysis results

```{r plot GO analysis results}

dotplot(ego, showCategory = 50,  orderBy="GeneRatio") +
      ggtitle("Dotplot - enrich GO") + scale_color_continuous(low = "purple", high = "green") 

```

## Correlation between hypoxia and steroidogenesis

```{r correlation between hypoxia and steroid genes}

# "Cortex_2D_vs_Tissue"  "Cortex_3D_vs_Tissue"  "Medulla_2D_vs_Tissue" "Medulla_3D_vs_Tissue" "Cortex_3D_vs_2D"     
# "Medulla_3D_vs_2D"   

d9 <- final[["Cortex_2D_vs_Tissue"]]
hypo_2d <- d9[which(d9$external_gene_name %in% hypoxia$gene),]
hypo_2d <- hypo_2d %>% arrange(log2FoldChange)
hypo_2d <- hypo_2d[1:25,]

ster_2d <- d9[which(d9$geneid %in% steroid),]
#ster_2d <- ster_2d %>% arrange(log2FoldChange)

# merge them into one dataframe
temp7 <- cbind(hypo_2d$log2FoldChange, ster_2d$log2FoldChange) %>% as.data.frame()
colnames(temp7) <- c("Hypoxia", "Steroidogenesis")
temp7$group <- rep("2D", 25)

d10 <- final[["Cortex_3D_vs_Tissue"]]
hypo_3d <- d10[which(d10$external_gene_name %in% hypoxia$gene),]
hypo_3d <- hypo_3d %>% arrange(log2FoldChange)
hypo_3d <- hypo_3d[1:24,]

ster_3d <- d10[which(d10$geneid %in% steroid),]
#ster_3d <- ster_3d %>% arrange(log2FoldChange)

temp8 <- cbind(hypo_3d$log2FoldChange, ster_3d$log2FoldChange) %>% as.data.frame()
colnames(temp8) <- c("Hypoxia", "Steroidogenesis")
temp8$group <- rep("3D", 24)

temp9 <- rbind(temp7, temp8)
ggplot(temp9, aes(x = Hypoxia, y = Steroidogenesis, col = group)) + geom_point() + theme + 
  xlab("Hypoxia log2FC") + ylab("Steroidogenesis log2FC") + 
  ggtitle("Cortex - Correlation between hypoxia and steroidogenesis") + 
  scale_color_manual(values = c("#f16c23","#1b7c3d"))

##### Medulla

d9 <- final[["Medulla_2D_vs_Tissue"]]
hypo_2d <- d9[which(d9$external_gene_name %in% hypoxia$gene),]
hypo_2d <- hypo_2d %>% arrange(log2FoldChange)

ster_2d <- d9[which(d9$geneid %in% steroid),]
hypo_2d <- hypo_2d[1:24,]

# merge them into one dataframe
temp7 <- cbind(hypo_2d$log2FoldChange, ster_2d$log2FoldChange) %>% as.data.frame()
colnames(temp7) <- c("Hypoxia", "Steroidogenesis")
temp7$group <- rep("2D", 24)

d10 <- final[["Medulla_3D_vs_Tissue"]]
hypo_3d <- d10[which(d10$external_gene_name %in% hypoxia$gene),]
hypo_3d <- hypo_3d %>% arrange(log2FoldChange)

ster_3d <- d10[which(d10$geneid %in% steroid),]
hypo_3d <- hypo_3d[1:25,]

temp8 <- cbind(hypo_3d$log2FoldChange, ster_3d$log2FoldChange) %>% as.data.frame()
colnames(temp8) <- c("Hypoxia", "Steroidogenesis")
temp8$group <- rep("3D", 25)

temp9 <- rbind(temp7, temp8)
ggplot(temp9, aes(x = Hypoxia, y = Steroidogenesis, col = group)) + geom_point() + theme + 
  xlab("Hypoxia log2FC") + ylab("Steroidogenesis log2FC") + 
  ggtitle("Medulla - Correlation between hypoxia and steroidogenesis") + 
  scale_color_manual(values = c("#f16c23","#1b7c3d"))

```

## Density plot for all genes

```{r density plot}

# get normalized count matrix for all expressed genes
cortex <- normalized_counts[[1]] 
cortex <- cortex[!duplicated(cortex$external_gene_name),]

scalecor <- scale(t(cortex[,2:(ncol(cortex)-3)]), scale = T, center = T) %>% t() %>% 
  as.data.frame() 

d2 <- cor_df

# Add up each column to get more row
temp5 <- scalecor[,1] %>% as.data.frame()
temp5$sample <- rep(colnames(scalecor[1]), nrow(temp5))

for (i in 2:ncol(scalecor)) {
  temp6 <- scalecor[,i] %>% as.data.frame()
  temp6$sample <- rep(colnames(scalecor[i]), nrow(temp6))
  temp5 <- rbind(temp5, temp6)
}

# Add the group information to the data frame
temp5$group <- NA
for (i in unique(temp5$sample)) {
  group <- d2[which(d2$name == i), which(colnames(d2) == "culture")]
  temp5[which(temp5$sample == i),3] <- rep(group, length(which(temp5$sample == i)))
}

ggplot(temp5, aes(x = ., col = group)) + geom_density() + theme +
  ggtitle("Density plot for normalized reads of all expressed genes")

# Kolmogorov-Smirnov Tests to test difference between distributions
dgof::ks.test(temp5$.[which(temp5$group %in% c("2D"))], 
              temp5$.[which(temp5$group %in% c("Tissue"))])

## Select genes with z-score between -2 to -1.5
temp6 <- temp5 %>% filter(z_score > -2) %>% filter(z_score < -1.5)

# Medulla
medulla <- normalized_counts[[2]] 
medulla <- medulla[!duplicated(medulla$external_gene_name),]
scalemed <- scale(t(medulla[,2:(ncol(medulla)-3)]), scale = T, center = T) %>% t() %>% 
  as.data.frame()

d2 <- med_df

# Add up each column to get more row
temp5 <- scalemed[,1] %>% as.data.frame()
temp5$sample <- rep(colnames(scalemed[1]), nrow(temp5))

for (i in 2:ncol(scalemed)) {
  temp6 <- scalemed[,i] %>% as.data.frame()
  temp6$sample <- rep(colnames(scalemed[i]), nrow(temp6))
  temp5 <- rbind(temp5, temp6)
}

# Add the group information to the data frame
temp5$group <- NA
for (i in unique(temp5$sample)) {
  group <- d2[which(d2$name == i), which(colnames(d2) == "culture")]
  temp5[which(temp5$sample == i),3] <- rep(group, length(which(temp5$sample == i)))
}

ggplot(temp5, aes(x = ., col = group)) + geom_density() + theme +
  ggtitle("Medulla - Density plot for normalized reads of all expressed genes")

```

## Proteomics correlation with transcriptomics

```{r proteomics correlation}

protein <- readxl::read_excel("/Users/tianyili/Desktop/Ovarioid/data/20220517_HF1_Valentina_BioSilk_Ovaroid_3D.xlsx")

description <- str_split(protein$Description, "=", simplify = T) %>% as.data.frame()
description$V4 <- gsub(" PE", "", description$V4)

protein$gene <- description$V4
pro <- protein %>% select(Description, gene, `Abundance: F10: Sample, Cortex`,
                          `Abundance: F11: Sample, Cortex`, `Abundance: F12: Sample, Cortex`,
                          `Abundance: F13: Sample, Cortex`, `Abundance: F6: Sample, Medulla`,
                          `Abundance: F7: Sample, Medulla`, `Abundance: F8: Sample, Medulla`,
                          `Abundance: F9: Sample, Medulla`)

angio_pro <- intersect(pro$gene, angio)
angio_protein <- pro[which(pro$gene %in% angio_pro),]

# Cortex
angio_protein_c <- angio_protein[, 3:6]
rownames(angio_protein_c) <- angio_protein$gene
scaleangio_c <- scale(t(angio_protein_c), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% 
  na.omit()

angio_c <- apply(scaleangio_c, 1, mean) %>% as.data.frame() %>% na.omit()
angio_c$gene <- rownames(angio_c)

df <- normalized_counts[["Cortex"]]
id_c <- match(angio_c$gene, df$external_gene_name) %>% na.omit()
d1 <- df[id_c,] 

genes_c <- d1[,(ncol(d1)-2)]

d2 <- cor_df %>% filter(culture == "3D")
d3 <- d1[, c(1, 16:18, which(colnames(d1) %in% d2$name))]

scalecor <- scale(t(d3[,5:(ncol(d3))]), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_c) %>% 
  na.omit()
rownames(scalecor) <- scalecor$external_gene_name

c3D <- scalecor[, -6]
c3DD <- apply(tis, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(c3D)) %>% 
  'colnames<-' ("transcript")

c3DD$gene <- rownames(c3DD)
c3DD_pro <- merge(angio_c, c3DD, by.x = "gene", by.y = "gene", all.x =T, all.y = F)
colnames(c3DD_pro) <- c("gene", "protein", "transcript")

rownames(c3DD_pro) <- c3DD_pro$gene
Heatmap(as.matrix(c3DD_pro[, 2:3]),
        cluster_columns = T,
        clustering_distance_rows = "euclidean",
        color = colorRampPalette(c("blue", "white", "red"))(100))

ggplot(c3DD_pro, aes(x = transcript, y = protein)) + geom_point() + theme +
  ggtitle("Angiogenesis - Cortex correlation")

```

## DWLS deconvolution

```{r DWLS deconvolution}

setwd("~/Desktop/CIIR/codes/DWLS/example/")

source("../Deconvolution_functions.R")

#get gene description
mart <- useEnsembl(biomart = "ensembl",dataset = "hsapiens_gene_ensembl")
output <- getBM(attributes=c("ensembl_gene_id","external_gene_name","entrezgene_id","description"), mart = mart)

# Load bulk data
bulk <- counts(dataset[["dds_Cortex"]], normalized = F) %>% as.data.frame()
bulk <- counts(dataset[["dds_Medulla"]], normalized = F) %>% as.data.frame()
bulk <- edgeR::cpm(bulk, normalized.lib.sizes = TRUE) %>% as.data.frame()
bulk$ENSEMBL <- rownames(bulk)
temp1 <- output[which(output$ensembl_gene_id %in% bulk$ENSEMBL),]
temp1 <- temp1[,1:2] 
colnames(temp1)[1] <- "ENSEMBL"

# Keep the genes with the gene names
bulk <- bulk[which(bulk$ENSEMBL %in% temp1$ENSEMBL),]
bulk_df <- base::merge(bulk, temp1, by.x = "ENSEMBL", by.y = "ENSEMBL", all.x = T, all.y = F)
bulk_df <- bulk_df[!duplicated(bulk_df),]
bulk_df <- bulk_df[-which(duplicated(bulk_df$external_gene_name)),]
bulk_df <- bulk_df[,-1]

# Relocate gene column, gene name must be in the first column of the bulk data matrix
bulk_df <- bulk_df %>% relocate(!!"external_gene_name")
colnames(bulk_df)[1] <- "gene_id"

# Load scRNAseq data
#sc <- readRDS("~/Desktop/PhD/Ovarioid/data/Unsorted_Manuscript figures.rds")
sc_in <- readRDS("~/Desktop/PhD/Ovarioid/data/Integrated_Manuscript figures.rds")

# Check single cell labels
#Idents(sc) <- sc@active.ident
#sc <- AddMetaData(sc, sc@active.ident, col.name = "clusters")
#sc_v5 <- UpdateSeuratObject(sc)

#DefaultAssay(sc_v5) <- "RNA"
#sc_sub <- subset(x = sc_v5, idents = c("2_immune", "3_gran", "4_endo", "5_pv", "6_stroma"))

# For integrated data
Idents(sc_in)
sc_in_v5 <- UpdateSeuratObject(sc_in)
sc_in_v5 <- AddMetaData(sc_in_v5, Idents(sc_in), col.name = "clusters")
DimPlot(sc_in_v5, group.by = "clusters", repel = T, label = T)
sc_in_v5 <- RenameIdents(sc_in_v5, "0" = "Stroma_theca", "1" = "Stroma_theca", "2" = "Stroma_theca", "9" = "Stroma_theca",
                         "13" = "Perivascular", "6" = "Perivascular", "3" = "Endothelial",
                         "12" = "Endothelial", "7" = "T_NK", "11" = "Granulosa",
                         "10" = "Monocytes", "4" = "Granulosa", "5" = "Granulosa", "8" = "Granulosa")
sc_in_v5 <- AddMetaData(sc_in_v5, sc_in_v5@active.ident, col.name = "integrated_clusters")
DimPlot(sc_in_v5, group.by = "integrated_clusters", repel = T, label = T)

DefaultAssay(sc_in_v5) <- "RNA"
#sc_sub <- subset(x = sc_in, idents = c("2_immune", "3_gran", "4_endo", "5_pv", "6_stroma"))
sc_sub <- sc_in_v5

# Get scRNAseq matrix
sc.mtx <- sc_sub[["RNA"]]$counts 
sc.meta <- sc_sub@meta.data %>% as.data.frame()

# Randomly sample 1000 cells in each group from each cluster. 
# If cluster has cells < 1000, get the cell id
cellid <- ""
gro <- unique(sc.meta$integrated_clusters)

for (i in gro) {
  temp1 <- sc.meta %>% filter(integrated_clusters == i)
  if (nrow(temp1) > 1000) {
    inTest <- sample(1:nrow(temp1), 1000, replace = F)
    id <- rownames(temp1)[inTest]
  } else if (nrow(temp1) < 1000) {
    id <- rownames(temp1)
  }
  cellid <- c(cellid, id)
}

# Subset the matrix, rownames of the sc.mtx should be the gene name
sc.mtx <- sc.mtx[, which(colnames(sc.mtx) %in% cellid)] %>% as.matrix()
sc.meta$cellid <- rownames(sc.meta)
sc.meta <- sc.meta[which(sc.meta$cellid %in% cellid),]

# Check the order and cell id are the same in labels and matrix
all(sc.meta$cellid == colnames(sc.mtx))
labels <- sc.meta$integrated_clusters

# Deconvolution
Signature <- buildSignatureMatrixMAST(sc.mtx, labels, "results")

allCounts_DWLS <- NULL
#allCounts_OLS <- NULL
#allCounts_SVR <- NULL
for (j in 1:(dim(bulk_df)[2]-1)) {
  S <- Signature
  Bulk <- bulk_df[,j+1]
  names(Bulk) <- bulk_df[,1]
  Genes <- intersect(rownames(S), names(Bulk))
  B <- Bulk[Genes]
  S <- S[Genes,]
  #solOLS <- solveOLS(S,B)
  solDWLS <- solveDampenedWLS(S,B)
  #solSVR <- solveSVR(S,B)
  
  allCounts_DWLS <- cbind(allCounts_DWLS, solDWLS)
  #allCounts_OLS <- cbind(allCounts_OLS, solOLS)
  #allCounts_SVR <- cbind(allCounts_SVR, solSVR)
}

# Assign the column name to the final result matrix
colnames(allCounts_DWLS) <- colnames(bulk_df)[-1]
df <- as.data.frame(allCounts_DWLS)
df$celltype <- rownames(df)
df <- df %>% tidyr::pivot_longer(!celltype, names_to = "sample", values_to = "percent")  

info <- read.csv("/Users/tili/Desktop/PhD/Ovarioid/data/processed/info_Cortex.csv") %>% dplyr::select(name, culture, group)
info <- read.csv("/Users/tili/Desktop/PhD/Ovarioid/data/processed/info_Medulla.csv") %>% dplyr::select(name, culture, group)

anno <- rbind(info, info, info, info, info, info)

colnames(df)[2] <- "name"
df1 <- cbind(df, anno)
colnames(df1)[4] <- "sample"
all(df1$name == df1$sample)

ggplot(df1, aes(x = name, y = percent, fill = celltype)) + geom_col(position = "stack") + 
  facet_grid(~culture, scales = "free") +
  theme_bw() + #facet_zoom(ylim = c(0.98, 1)) + 
  xlab("") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_manual(values = c("Stroma_theca" = "grey", "Granulosa" = "#00B6EB", "T_NK" = "#A58AFF",
                               "Monocytes" = "#f4a582",
                               "Endothelial" = "#53B400", "Perivascular" = "#C49A00"))

df1 <- df1 %>% dplyr::filter(culture != "Tissue")
df1 %>% dplyr::filter(culture == "3D") %>% 
  ggplot(aes(x = name, y = percent, fill = celltype)) + geom_col(position = "stack") + #facet_grid(~culture) +
  theme_bw() + facet_zoom(ylim = c(0.97, 1)) + xlab("") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_manual(values = c("Stroma_theca" = "grey", "Granulosa" = "#00B6EB", "T_NK" = "#A58AFF",
                               "Monocytes" = "#f4a582",
                               "Endothelial" = "#53B400", "Perivascular" = "#C49A00"))

df1 %>% dplyr::filter(culture == "2D") %>% 
  ggplot(aes(x = name, y = percent, fill = celltype)) + geom_col(position = "stack") + #facet_grid(~culture) +
  theme_bw() + facet_zoom(ylim = c(0.98, 1)) + xlab("") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_manual(values = c("Stroma_theca" = "grey", "Granulosa" = "#00B6EB", "T_NK" = "#A58AFF",
                               "Monocytes" = "#f4a582",
                               "Endothelial" = "#53B400", "Perivascular" = "#C49A00"))

```

## Figure 6b: Plot individual genes

```{r individual gene plotting}

col1a1 <- plotCounts(dataset[["dds_Medulla"]], gene= "ENSG00000108821", intgroup=c("freezing", "culture"), normalized = T,
                  returnData=TRUE) 
col1a1$gene <- rep("COL1A1", nrow(col1a1))

lama1 <- plotCounts(dataset[["dds_Medulla"]], gene= "ENSG00000101680", intgroup=c("freezing", "culture"), normalized = T,
                  returnData=TRUE) 
lama1$gene <- rep("LAMA1", nrow(lama1))

d7 <- rbind(col1a1, lama1)
d7$culture <- factor(d7$culture, levels = c("Tissue", "2D", "3D"))

ggplot(d7, aes(x=culture, y=count, fill = culture)) + 
    geom_boxplot() + facet_wrap(~gene, scales = "free_y") +
    xlab("") + ylab("normalized count") +
    theme_bw() + 
    scale_fill_manual(values=c("#f0f0f0", "#b3cde3", "orange"))

```

## Figure 6a: Selected marker expression

```{r selected markers}

cort <- readRDS("/Users/tili/Desktop/PhD/POPs/data/Unsorted_Manuscript figures.rds")
cort = UpdateSeuratObject(object = cort)

markers <- c("COL1A1", "FN1", "COL1A2", "COL3A1", "COL4A2", "COL4A1", "COL5A1", "COL5A2",
             "ITGB1", "MMP2", "MMP14", "BGN", "VCAN", "ITGA11", "MGP", "FBN1", "COL15A1",
             "PXDN", "ITGB5", "LAMC1", "VCL", "COL8A1", "LAMB1", "ITGA3", "ELN", "ITGA5", "LAMA1")

cort <- RenameIdents(cort, "1_oocytes" = "Oocytes", "2_immune" = "Immune", "3_gran" = "Granulosa", "4_endo" = "Endothelial", "5_pv" = "Perivascular", "6_stroma" = "Stroma")

DotPlot(cort, features = markers) + coord_flip() + xlab("") + ylab("")

```

## Selected markers in RNAseq - 2D removed

```{r cytokine - 2D removed}

cyto <- c("ENSG00000136244", "ENSG00000169429", "ENSG00000108691", "ENSG00000107562", "ENSG00000275302",
          "ENSG00000172156", "ENSG00000277632", "ENSG00000136689", "ENSG00000110944", "ENSG00000164400", 
          "ENSG00000136634")

# Cortex cytokine
df <- normalized_counts[["Cortex"]]
id_c <- match(cyto, df$gene) %>% na.omit()
d1 <- df[id_c,] 
d1 <- na.omit(d1) %>% as.data.frame()
rownames(d1) <- d1$external_gene_name
d1 <- d1[!duplicated(d1$external_gene_name),]
d2 <- cor_df %>% filter(culture != "2D")
genes_c <- d1[,(ncol(d1)-2)]
d1 <- d1[,which(colnames(d1) %in% d2$name)]

scalecor <- scale(t(d1), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_c) %>% 
  na.omit()
rownames(scalecor) <- scalecor$genes_c

d2 <- d2[which(d2$name %in% colnames(scalecor)),]

anno <- d2[, c(1, 2, 7, 11)]
rownames(anno) <- anno$name

pheatmap(scalecor[, -ncol(scalecor)], cluster_rows = F, show_rownames = T, 
         cluster_cols = T, clustering_distance_rows = "correlation", annotation = anno, 
         color = colorRampPalette(c("blue", "white", "red"))(100),
         breaks = seq(-1, 1, length.out = 90))

d4 <- anno %>% filter(culture == "Tissue") #%>% filter(type == "Cortex")
tis <- scalecor[,which(colnames(scalecor) %in% d4$name)]
tiss <- apply(tis, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(tis)) %>% 
  'colnames<-' ("Tissue")

d6 <- anno %>% filter(culture == "3D") #%>% filter(type == "Cortex")
c3D <- scalecor[,which(colnames(scalecor) %in% d6$name)]
c3DD <- apply(c3D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(c3D)) %>% 
  'colnames<-' ("3D")


# Medulla cytokine
df <- normalized_counts[["Medulla"]]
id_m <- match(cyto, df$gene) %>% na.omit()
d1 <- df[id_m,] 
d1 <- na.omit(d1) %>% as.data.frame()
rownames(d1) <- d1$external_gene_name
d1 <- d1[!duplicated(d1$external_gene_name),]
d2 <- med_df %>% filter(culture != "2D")

genes_m <- d1[,(ncol(d1)-2)]
d1 <- d1[,which(colnames(d1) %in% d2$name)]

scalemed <- scale(t(d1), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_m) %>% 
  na.omit()
rownames(scalemed) <- scalemed$genes_m

d2 <- d2[which(d2$name %in% colnames(scalemed)),]

anno <- d2[, c(1, 2, 7, 11)]
rownames(anno) <- anno$name

pheatmap(scalemed[, -ncol(scalemed)], cluster_rows = F, show_rownames = T, 
         cluster_cols = T, clustering_distance_rows = "correlation", annotation = anno, 
         color = colorRampPalette(c("blue", "white", "red"))(100),
         breaks = seq(-1, 1, length.out = 100))

d7 <- anno %>% filter(culture == "Tissue") #%>% filter(type == "Medulla")
mtis <- scalemed[,which(colnames(scalemed) %in% d7$name)]
mtiss <- apply(mtis, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(mtis)) %>% 
  'colnames<-' ("Tissue_med")

d9 <- anno %>% filter(culture == "3D") #%>% filter(type == "Medulla")
m3D <- scalemed[,which(colnames(scalemed) %in% d9$name)]
m3DD <- apply(m3D, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(m3D)) %>% 
  'colnames<-' ("3D_med")

# Combine cortex and medulla but keep the z score
con <- cbind(tiss, c3DD, mtiss, m3DD)

anno_group <- colnames(con) %>% as.data.frame()
anno_group$group <- rep(c("Cortex", "Medulla"), each = 2)

Heatmap(as.matrix(con), name = "Z-score",
        cluster_columns = F,
        clustering_distance_rows = "euclidean",
        show_row_names = T,
        column_split = anno_group$group,
        color = colorRampPalette(c("blue", "white", "red"))(100))

```



## Hypoxia genes in IVA dataset

```{r IVA dataset hypoxia}

read_excel_allsheets <- function(filename, tibble = FALSE) {
    sheets <- readxl::excel_sheets(filename)
    x <- lapply(sheets, function(X) readxl::read_excel(filename, sheet = X))
    if(!tibble) x <- lapply(x, as.data.frame)
    names(x) <- sheets
    x
}

kgn <- read_excel_allsheets("/Users/tianyili/Desktop/Ovarioid/results/hypoxia_genes.xlsx")
hypoxia <- kgn[[1]]

dds <- DESeqDataSetFromMatrix(countData = iva_bulk, colData = iva_info, design = ~cultures)
keep <- rowSums(counts(dds) > 1) >= 7
dds <- dds[keep,]
dds <- DESeq(dds)

normalized_counts <- counts(dds, normalized=T) %>% 
  data.frame() %>%
  tibble::rownames_to_column(var="gene") %>%
  as_tibble() %>%
  left_join(output, by=c("gene" = "ensembl_gene_id")) 

# Cortex hypoxia
df <- normalized_counts
id_c <- match(hypoxia$gene, df$external_gene_name) %>% na.omit()
d1 <- df[id_c,] 
d1 <- na.omit(d1)

genes_c <- d1[,(ncol(d1)-2)]

scalecor <- scale(t(d1[,2:(ncol(d1)-3)]), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_c) %>% 
  na.omit()
rownames(scalecor) <- scalecor$external_gene_name

d2 <- iva_info

anno <- d2[, c(1, 6)]
rownames(anno) <- anno$Sample

pheatmap(scalecor[, -ncol(scalecor)], cluster_rows = T, show_rownames = T, 
         cluster_cols = T, clustering_distance_rows = "correlation", annotation = anno, 
         color = colorRampPalette(c("blue", "white", "red"))(100),
         breaks = seq(-1, 1, length.out = 90))

d4 <- anno %>% filter(culture == "Fresh")
tis <- scalecor[,which(colnames(scalecor) %in% d4$Sample)]
tiss <- apply(tis, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(tis)) %>% 
  'colnames<-' ("Tissue")

d5 <- anno %>% filter(culture == "Cultured")
frag <- scalecor[,which(colnames(scalecor) %in% d5$Sample)]
fragg <- apply(frag, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(frag)) %>% 
  'colnames<-' ("Fragmentation")

cortex <- cbind(tiss, fragg)

anno_group <- colnames(cortex) %>% as.data.frame()
colnames(anno_group) <- "group"

Heatmap(as.matrix(cortex), name = "Z-score",
        #cluster_row_slices = T,
        cluster_columns = F,
        clustering_distance_rows = "euclidean",
        row_km = 2,
        #row_gap = unit(2, "mm"),
        show_row_names = T,
        column_split = anno_group$group,
        color = colorRampPalette(c("blue", "white", "red"))(100))

```

## IVA angiogenesis heatmap

```{r IVA angiogenesis}

# Cortex angiogenesis

angiogenesis <- bitr("GO:0001525", fromType = "GO", toType = "SYMBOL", OrgDb = org.Hs.eg.db)
angio <- angiogenesis$SYMBOL

df <- normalized_counts
id_c <- match(angio, df$external_gene_name) %>% na.omit()
d1 <- df[id_c,] 
d1 <- na.omit(d1)
d1 <- d1[!duplicated(d1$external_gene_name),]

genes_c <- d1[,(ncol(d1)-2)]

scalecor <- scale(t(d1[,2:(ncol(d1)-3)]), scale = T, center = T) %>% t() %>% 
  as.data.frame() %>% cbind(., genes_c) %>% 
  na.omit()
rownames(scalecor) <- scalecor$external_gene_name

d2 <- iva_info

anno <- d2[, c(1, 6)]
rownames(anno) <- anno$Sample

pheatmap(scalecor[, -ncol(scalecor)], cluster_rows = T, show_rownames = T, 
         cluster_cols = T, clustering_distance_rows = "correlation", annotation = anno, 
         color = colorRampPalette(c("blue", "white", "red"))(100),
         breaks = seq(-1, 1, length.out = 90))

d4 <- anno %>% filter(culture == "Fresh")
tis <- scalecor[,which(colnames(scalecor) %in% d4$Sample)]
tiss <- apply(tis, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(tis)) %>% 
  'colnames<-' ("Tissue")

d5 <- anno %>% filter(culture == "Cultured")
frag <- scalecor[,which(colnames(scalecor) %in% d5$Sample)]
fragg <- apply(frag, 1, mean) %>% as.data.frame() %>% 
  'rownames<-' (rownames(frag)) %>% 
  'colnames<-' ("Fragmentation")

cortex <- cbind(tiss, fragg)

anno_group <- colnames(cortex) %>% as.data.frame()
colnames(anno_group) <- "group"

Heatmap(as.matrix(cortex), name = "Z-score",
        #cluster_row_slices = T,
        cluster_columns = F,
        clustering_distance_rows = "euclidean",
        row_km = 2,
        #row_gap = unit(2, "mm"),
        show_row_names = T,
        column_split = anno_group$group,
        color = colorRampPalette(c("blue", "white", "red"))(100))


```

## IVA - Density plot for all genes

```{r IVA density plot}

# get normalized count matrix for all expressed genes
cortex <- normalized_counts
scalecor <- scale(t(cortex[,2:(ncol(cortex)-3)]), scale = T, center = T) %>% t() %>% 
  as.data.frame()

d2 <- iva_info

# Add up each column to get more row
temp5 <- scalecor[,1] %>% as.data.frame()
temp5$sample <- rep(colnames(scalecor[1]), nrow(temp5))

for (i in 2:ncol(scalecor)) {
  temp6 <- scalecor[,i] %>% as.data.frame()
  temp6$sample <- rep(colnames(scalecor[i]), nrow(temp6))
  temp5 <- rbind(temp5, temp6)
}

# Add the group information to the data frame
temp5$group <- NA
for (i in unique(temp5$sample)) {
  group <- d2[which(d2$Sample == i), which(colnames(d2) == "culture")]
  temp5[which(temp5$sample == i),3] <- rep(group, length(which(temp5$sample == i)))
}

ggplot(temp5, aes(x = ., col = group)) + geom_density() + theme_bw() +
  ggtitle("Density plot - IVA") + xlab("Z-score")

```

## DE analysis between tissue cortex and medulla

```{r DE analysis cortex medulla}

cor_info <- read.csv("/Users/tianyili/Desktop/Ovarioid/data/processed/info_Cortex.csv", sep = ";")
cor_cells <- read.delim("/Users/tianyili/Desktop/Ovarioid/data/processed/cells_Cortex.csv", sep = ",")
cor_cells <- re_name(cor_cells)
print(all(cor_info$name == colnames(cor_cells)))

med_info <- read.csv("/Users/tianyili/Desktop/Ovarioid/data/processed/info_Medulla.csv", sep = ";")
med_cells <- read.delim("/Users/tianyili/Desktop/Ovarioid/data/processed/cells_Medulla.csv", sep = ",")
med_cells <- re_name(med_cells)
print(all(med_info$name == colnames(med_cells)))

al <- cbind(cor_cells, med_cells)
al_info <- rbind(cor_info, med_info)
al_info <- al_info %>% filter(culture == "Tissue")
al <- re_order(al_info, al)

dds <- DESeqDataSetFromMatrix(countData=al, colData = al_info, design = ~ type)
keep <- rowSums(counts(dds) > 1) >= 5
dds <- dds[keep,]

dds <- DESeq(dds)

comALL <- results(dds, contrast = c("type", "Medulla", "Cortex"), 
                                    independentFiltering = T, pAdjustMethod = "fdr") %>% 
                                    as.data.frame() %>% 
                                    mutate(geneid = rownames(.))
com05 <- comALL %>% filter(padj < 0.05)

final05 <- get_id(com05)
finalALL <- get_id(comALL)

openxlsx::write.xlsx(final05, file = "/Users/tianyili/Desktop/Ovarioid/results/tables/DESeq2_cortex_vs_medulla_FDR_0.05.xlsx")

edo <- enrichGO(gene = final05$entrezgene_id,
                universe = finalALL$entrezgene_id,
                OrgDb = org.Hs.eg.db, ont = "ALL", 
                pAdjustMethod = "BH", 
                pvalueCutoff = 0.05, readable = TRUE)
edo <- setReadable(edo, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
res <- edo@result %>% arrange(p.adjust)

openxlsx::write.xlsx(res, file = "/Users/tianyili/Desktop/Ovarioid/results/tables/DESeq2_cortex_vs_medulla_enrichGO_FDR_0.05.xlsx")

```

## Session info

```{r session info}

sessionInfo()

```

