---
title: "Ovarioid_proteomics_analysis"
output: html_document
date: "2023-11-10"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load required library

```{r Load dependencies}
suppressMessages(library(ggplot2))
suppressMessages(library(pheatmap))
suppressMessages(library(openxlsx))
suppressMessages(library(readxl))
suppressMessages(library(biomaRt))
suppressMessages(library(org.Hs.eg.db))
suppressMessages(library(DOSE))
suppressMessages(library(pathview))
suppressMessages(library(clusterProfiler))
suppressMessages(library(pheatmap))
suppressMessages(library(VennDiagram))
suppressMessages(library(RColorBrewer))
suppressMessages(library(msigdbr))
suppressMessages(library(edgeR))
suppressMessages(library(stringr))
suppressMessages(library(future))
suppressMessages(library(patchwork))
suppressMessages(library(VennDiagram))
suppressMessages(library(GSEABase))
suppressMessages(library(GSVA))
library(Seurat)
library(tidyverse)
```

## Import data

```{r import data, echo=FALSE}

tissue <- read_excel("/Users/tianyili/Desktop/Ovarioid/data/proteomics/20230929_HF2_01_AS1_Valentina_tissue_TMT6.xlsx")
bso <- read_excel("/Users/tianyili/Desktop/Ovarioid/data/proteomics/20230929_HF2_02_AS1_Valentina_BSO_TMT6.xlsx")

tis <- tissue[, c(24, 26, 36:41)]
so <- bso[, c(24, 26, 36:41)]

# Keep the common protein only between tissue and bso
universe <- intersect(tis$`Gene Symbol`, so$`Gene Symbol`)
tis <- tis[which(tis$`Gene Symbol` %in% universe),]
so <- so[which(so$`Gene Symbol` %in% universe),]

# change colume name
colnames(tis) <- c("Gene", "WikiPathways", "Tissue_127N_cortex", "Tissue_129N_cortex", "Tissue_130N_cortex",
                   "Tissue_126_medulla", "Tissue_128N_medulla", "Tissue_131_medulla")
colnames(so) <- c("Gene", "WikiPathways", "BSO_127N_cortex", "BSO_129N_cortex", "BSO_130N_cortex",
                   "BSO_126_medulla", "BSO_128N_medulla", "BSO_131_medulla")

# Merge cortex samples from tissue and BSO, change NA to 0
cortex <- merge(tis[, 1:5], so[, 1:5], by.x = "Gene", by.y = "Gene")
cortex <- cortex[, -6]
cortex[is.na(cortex)] <- 0

medulla <- merge(tis[, c(1:2, 6:8)], so[, c(1:2, 6:8)], by.x = "Gene", by.y = "Gene")
medulla <- medulla[, -6]
medulla[is.na(medulla)] <- 0

```

## Normalization 

```{r normalization}

# Within each sample, divide the level of each single protein by the sum of all the proteins
cor_sum <- apply(cortex[, 3:8], 2, sum)
med_sum <- apply(medulla[, 3:8], 2, sum)

cor <- cortex[, 1:2]
cor$T127 <- cortex$Tissue_127N_cortex / 2244580.3
cor$T129 <- cortex$Tissue_129N_cortex / 926404.0
cor$T130 <- cortex$Tissue_130N_cortex / 1003005.3
cor$B127 <- cortex$BSO_127N_cortex / 309364.0
cor$B129 <- cortex$BSO_129N_cortex / 156567.2
cor$B130 <- cortex$BSO_130N_cortex / 2095.7

med <- medulla[, 1:2]
med$T126 <- medulla$Tissue_126_medulla / 2151360.7
med$T128 <- medulla$Tissue_128N_medulla / 1935841.4
med$T131 <- medulla$Tissue_131_medulla / 1458194.0
med$B126 <- medulla$BSO_126_medulla / 384144.7
med$B128 <- medulla$BSO_128N_medulla / 1655849.8
med$B131 <- medulla$BSO_131_medulla / 481002.2

```

## Angiogenesis gene plotting

```{r angiogenesis genes}

angiogenesis <- bitr("GO:0001525", fromType = "GO", toType = "SYMBOL", OrgDb = org.Hs.eg.db)
angio <- angiogenesis$SYMBOL

whole <- cbind(cor, med)
# Remove outlier and duplicated columns
whole <- whole[, -c(8:10)]

angio_cor <- whole[which(whole$Gene %in% angio),]
rownames(angio_cor) <- angio_cor$Gene
angio_cor <- angio_cor[, -2]

## Individual plot
angio_cort <- angio_cor %>% 
  tidyr::pivot_longer(!Gene, names_to = "sample", values_to = "contribution")

ggplot(angio_cort, aes(x = sample, y = Gene, fill = log2(contribution + 1))) + geom_tile() +
  scale_fill_distiller()

angio_cort$group <- ifelse(str_detect(angio_cort$sample, "T"), "Tissue", "BSO")
angio_cort$group <- factor(angio_cort$group, levels = c("Tissue", "BSO"))
angio_cort %>% filter(Gene == "TGFBI") %>% 
  ggplot(aes(x = group, y = contribution)) + geom_boxplot() + theme_bw() +
  ggtitle("TGFBI")

## Heatmap
new_cols <- c("Tissue_Cortex", "Silk_Ova_Cortex", "Tissue_Medulla", "Silk_Ova_Medulla")
angio_cor <- angio_cor[,-1]
# Get the average expression among three biological replicates
df <- sapply(seq(1, ncol(angio_cor), 3), function(j) {
  if (j < 4) {
    # if keep 10x group, then j < 43
    base::rowMeans(angio_cor[, j+(0:2)])
  } else if (j == 4) { 
    base::rowMeans(angio_cor[, j+(0:1)]) } else {
      base::rowMeans(angio_cor[, j-1+(0:2)])
    }
  }) %>% 
  as.data.frame() %>% 
  rlang::set_names(new_cols)

anno <- colnames(df) %>% as.data.frame() 
anno$type <- rep(c("Cortex", "Medulla"), each = 2) 
anno$culture <- rep(c("Tissue", "Silk-Ova"), 2)
rownames(anno) <- anno$.
anno <- anno[,-1]
anno$type <- factor(anno$type, levels = c("Cortex", "Medulla"))
anno$culture <- factor(anno$culture, levels = c("Tissue", "Silk-Ova"))

scalecor <- scale(t(df), scale = T, center = T) %>% t() %>% 
  as.data.frame()

Heatmap(as.matrix(scalecor), name = "Z-score",
        cluster_rows = T,
        show_row_names = T,
        cluster_columns = F,
        clustering_distance_rows = "euclidean",
        column_split = anno$type,
        color = colorRampPalette(c("blue", "white", "red"))(100))

```

## Hypoxia gene plotting

```{r hypoxia genes}

hypoxia <- read.xlsx("/Users/tianyili/Desktop/Ovarioid/results/hypoxia_genes.xlsx")
hypo <- hypoxia$gene

whole <- cbind(cor, med)
# Remove outlier and duplicated columns
whole <- whole[, -c(8:10)]

hypoxia_cor <- whole[which(whole$Gene %in% hypo),]
rownames(hypoxia_cor) <- hypoxia_cor$Gene
hypoxia_cor <- hypoxia_cor[, -2]

## Heatmap
new_cols <- c("Tissue_Cortex", "Silk_Ova_Cortex", "Tissue_Medulla", "Silk_Ova_Medulla")
hypoxia_cor <- hypoxia_cor[,-1]
# Get the average expression among three biological replicates
df <- sapply(seq(1, ncol(hypoxia_cor), 3), function(j) {
  if (j < 4) {
    # if keep 10x group, then j < 43
    base::rowMeans(hypoxia_cor[, j+(0:2)])
  } else if (j == 4) { 
    base::rowMeans(hypoxia_cor[, j+(0:1)]) } else {
      base::rowMeans(hypoxia_cor[, j-1+(0:2)])
    }
  }) %>% 
  as.data.frame() %>% 
  set_names(new_cols)

anno <- colnames(df) %>% as.data.frame() 
anno$type <- rep(c("Cortex", "Medulla"), each = 2) 
anno$culture <- rep(c("Tissue", "Silk-Ova"), 2)
rownames(anno) <- anno$.
anno <- anno[,-1]
anno$type <- factor(anno$type, levels = c("Cortex", "Medulla"))
anno$culture <- factor(anno$culture, levels = c("Tissue", "Silk-Ova"))

scalecor <- scale(t(df), scale = T, center = T) %>% t() %>% 
  as.data.frame()

Heatmap(as.matrix(scalecor), name = "Z-score",
        cluster_rows = T,
        show_row_names = T,
        cluster_columns = F,
        clustering_distance_rows = "euclidean",
        column_split = anno$type,
        color = colorRampPalette(c("blue", "white", "red"))(100))

```

## correlation between transcriptomics and proteomics - Angiogenesis Cortex

```{r correlation cortex angiogenesis}

# Subset transcriptomics data to protein detected
cort <- normalized_counts[["Cortex"]]

# Merge proteomics data
angiogenesis <- bitr("GO:0001525", fromType = "GO", toType = "SYMBOL", OrgDb = org.Hs.eg.db)
angio <- angiogenesis$SYMBOL

whole <- cbind(cor, med)
# Remove outlier and duplicated columns
whole <- whole[, -c(8:10)]

angio_cor <- whole[which(whole$Gene %in% angio),]
rownames(angio_cor) <- angio_cor$Gene
angio_cor <- angio_cor[, -2]

# Subset protein dataset to cortex
angio_cortex <- angio_cor[,1:6]
#angio_cortex$Gene <- rownames(angio_cortex)
cort_df <- cort[which(cort$external_gene_name %in% angio_cortex$Gene),]

cortex <- merge(angio_cortex, cort_df, by.x = "Gene", by.y = "external_gene_name")
cortex <- cortex[, -c(7,22:23)]

# Read the metadata and keep only the samples that match the proteomics
info <- read.csv("/Users/tianyili/Desktop/Ovarioid/data/processed/info_Cortex.csv", sep = ";")
info_cor <- info %>% select(name, culture) %>% 
  filter(culture != "2D")
info_cor[11:15,1] <- colnames(angio_cortex)[2:6]
info_cor[11:15,2] <- c("Tissue", "Tissue", "Tissue", "3D", "3D")

# Remove samples not used for proteomics "P28012_1003_S3"  "P28012_1001_S1"  "P28012_1014_S14" "P28012_1011_S11"
dif <- c("P28012_1003_S3", "P28012_1001_S1", "P28012_1014_S14", "P28012_1011_S11")
info_cor <- info_cor[-which(info_cor$name %in% dif),]
# Remove the outlier sample
info_cor <- info_cor[-which(info_cor$name == "P28012_1043_S43"),]
info_cor <- info_cor %>% arrange(culture)

rownames(cortex) <- cortex$Gene
cortex <- cortex[,which(colnames(cortex) %in% info_cor$name)]
cortex$Gene <- rownames(cortex)

cor_protein <- cortex[,c(1:5,11)] %>% 
  tidyr::pivot_longer(!Gene, names_to = "sample", values_to = "contribution")

cor_RNA <- cortex[, 6:11]
cor_RNA <- cor_RNA[,match(colnames(cor_RNA), c("P28012_1022_S22", "P28012_1053_S53", "P28012_1038_S38",
                           "P28012_1027_S27", "P28012_1058_S58", "Gene"))]
cortex_RNA <- cor_RNA
cortex_RNA <- cortex_RNA %>% 
  tidyr::pivot_longer(!Gene, names_to = "sample", values_to = "contribution")
colnames(cortex_RNA) <- paste0(colnames(cortex_RNA), "_RNA")

cort <- cbind(cortex_RNA, cor_protein)
cort$type <- ifelse(str_detect(cort$sample, "T"), "Tissue", "Silk-Ov")
cort <- cort %>% filter(contribution > 0)

p1 <- ggplot(cort, aes(x = contribution_RNA, y = contribution, col = type)) + 
  geom_point(aes(shape = type)) + theme_bw() +
  geom_smooth(method = "lm", se = F) + scale_color_manual(values = c("#1B7C3D", "#2B6A99")) +
  ggtitle("Angiogenesis - Cortex correlation") + xlim(c(0, 500000))

# Correlation test
tis_cor <- cort %>% filter(type == "Tissue")

# Remove outlier
out <- which(tis_cor$contribution_RNA > 500000)
out # no outlier
cor.test(tis_cor$contribution_RNA, tis_cor$contribution, method = "pearson")

silk_cor <- cort %>% filter(type != "Tissue")

# Remove outlier
out <- which(silk_cor$contribution_RNA > 500000)
out # no outlier
cor.test(silk_cor$contribution_RNA[-out], silk_cor$contribution[-out], method = "pearson")

```

## correlation between transcriptomics and proteomics - Angiogenesis Medulla

```{r correlation medulla angiogenesis}

# Subset transcriptomics data to protein detected
medu <- normalized_counts[["Medulla"]]
medu_df <- medu[which(medu$external_gene_name %in% angio_cor$Gene),]

# Subset protein dataset to cortex
angio_medulla <- angio_cor[,c(1, 7:12)]

medulla <- merge(angio_medulla, medu_df, by.x = "Gene", by.y = "external_gene_name")
medulla <- medulla[, -c(8,24:25)]

# Read the metadata and keep only the samples that match the proteomics
info <- read.csv("/Users/tianyili/Desktop/Ovarioid/data/processed/info_Medulla.csv", sep = ";")
info_med <- info %>% select(name, culture) %>% 
  filter(culture != "2D")
info_med[11:16,1] <- colnames(angio_medulla)[2:7]
info_med[11:16,2] <- c("Tissue", "Tissue", "Tissue", "3D", "3D", "3D")

# Remove samples not used for proteomics "P28012_1008_S8", "P28012_1005_S5", "P28012_1019_S19", "P28012_1016_S16"
dif <- c("P28012_1008_S8", "P28012_1005_S5", "P28012_1019_S19", "P28012_1016_S16")
info_med <- info_med[-which(info_med$name %in% dif),]
info_med <- info_med %>% arrange(culture)

rownames(medulla) <- medulla$Gene
medulla <- medulla[,which(colnames(medulla) %in% info_med$name)]
medulla$Gene <- rownames(medulla)

med_protein <- medulla[,c(1:6,13)] %>% 
  tidyr::pivot_longer(!Gene, names_to = "sample", values_to = "contribution")

med_RNA <- medulla[, 7:13]
id <- match(colnames(med_RNA), c("P28012_1030_S30", "P28012_1046_S46", "P28012_1061_S61",
                           "P28012_1035_S35", "P28012_1051_S51", "P28012_1066_S66", "Gene"))
med_RNA <- med_RNA[, id]

medulla_RNA <- med_RNA
medulla_RNA <- medulla_RNA %>% 
  tidyr::pivot_longer(!Gene, names_to = "sample", values_to = "contribution")
colnames(medulla_RNA) <- paste0(colnames(medulla_RNA), "_RNA")

medul <- cbind(medulla_RNA, med_protein)
medul <- medul %>% filter(contribution > 0)
medul$type <- ifelse(str_detect(medul$sample, "T"), "Tissue", "Silk-Ov")

p2 <- ggplot(medul, aes(x = contribution_RNA, y = contribution, col = type)) + 
  geom_point(aes(shape = type)) + theme_bw() +
  geom_smooth(method = "lm", se = F) + scale_color_manual(values = c("#1B7C3D", "#2B6A99")) +
  ggtitle("Angiogenesis - Medulla correlation") + xlim(c(0, 500000))

# Correlation test
tis_med <- medul %>% filter(type == "Tissue")

# Remove outlier
out <- which(tis_med$contribution_RNA > 500000)
out # no outlier
cor.test(tis_med$contribution_RNA, tis_med$contribution, method = "pearson")

silk_med <- medul %>% filter(type != "Tissue")

# Remove outlier
out <- which(silk_med$contribution_RNA > 500000)
out # no outlier
cor.test(silk_med$contribution_RNA[-out], silk_med$contribution[-out], method = "pearson")

```

## correlation between transcriptomics and proteomics - Hypoxia Cortex

```{r correlation cortex hypoxia}

# Subset transcriptomics data to protein detected
cort <- normalized_counts[["Cortex"]]

hypoxia <- read.xlsx("/Users/tianyili/Desktop/Ovarioid/results/hypoxia_genes.xlsx")
hypo <- hypoxia$gene

whole <- cbind(cor, med)
# Remove outlier and duplicated columns
whole <- whole[, -c(8:10)]

hypoxia_cor <- whole[which(whole$Gene %in% hypo),]
rownames(hypoxia_cor) <- hypoxia_cor$Gene
hypoxia_cor <- hypoxia_cor[, -2]

cort_df <- cort[which(cort$external_gene_name %in% hypoxia_cor$Gene),]

# Subset protein dataset to cortex
hypoxia_cortex <- hypoxia_cor[,1:6]

cortex <- merge(hypoxia_cortex, cort_df, by.x = "Gene", by.y = "external_gene_name")
cortex <- cortex[, -c(7,22:23)]

# Read the metadata and keep only the samples that match the proteomics
info <- read.csv("/Users/tianyili/Desktop/Ovarioid/data/processed/info_Cortex.csv", sep = ";")
info_cor <- info %>% select(name, culture) %>% 
  filter(culture != "2D")
info_cor[11:15,1] <- colnames(angio_cortex)[2:6]
info_cor[11:15,2] <- c("Tissue", "Tissue", "Tissue", "3D", "3D")

# Remove samples not used for proteomics "P28012_1003_S3"  "P28012_1001_S1"  "P28012_1014_S14" "P28012_1011_S11"
dif <- c("P28012_1003_S3", "P28012_1001_S1", "P28012_1014_S14", "P28012_1011_S11")
info_cor <- info_cor[-which(info_cor$name %in% dif),]
# Remove the outlier sample
info_cor <- info_cor[-which(info_cor$name == "P28012_1043_S43"),]
info_cor <- info_cor %>% arrange(culture)

rownames(cortex) <- cortex$Gene
cortex <- cortex[,which(colnames(cortex) %in% info_cor$name)]
cortex$Gene <- rownames(cortex)

cor_protein <- cortex[,c(1:5,11)] %>% 
  tidyr::pivot_longer(!Gene, names_to = "sample", values_to = "contribution")

cor_RNA <- cortex[, 6:11]
cor_RNA <- cor_RNA[,match(colnames(cor_RNA), c("P28012_1022_S22", "P28012_1053_S53", "P28012_1038_S38",
                           "P28012_1027_S27", "P28012_1058_S58", "Gene"))]
cortex_RNA <- cor_RNA
cortex_RNA <- cortex_RNA %>% 
  tidyr::pivot_longer(!Gene, names_to = "sample", values_to = "contribution")
colnames(cortex_RNA) <- paste0(colnames(cortex_RNA), "_RNA")

cort <- cbind(cortex_RNA, cor_protein)
cort$type <- ifelse(str_detect(cort$sample, "T"), "Tissue", "Silk-Ov")
cort <- cort %>% filter(contribution > 0)

p3 <- ggplot(cort, aes(x = contribution_RNA, y = contribution, col = type)) + 
  geom_point(aes(shape = type)) + theme_bw() +
  geom_smooth(method = "lm", se = F) + scale_color_manual(values = c("#1B7C3D", "#2B6A99")) +
  ggtitle("Hypoxia - Cortex correlation") + xlim(c(0, 100000))

# Correlation test
tis_cor <- cort %>% filter(type == "Tissue")

# Remove outlier
out <- which(tis_cor$contribution_RNA > 100000)
out # no outlier
cor.test(tis_cor$contribution_RNA, tis_cor$contribution, method = "pearson")

silk_cor <- cort %>% filter(type != "Tissue")

# Remove outlier
out <- which(silk_cor$contribution_RNA > 100000)
out # no outlier
cor.test(silk_cor$contribution_RNA[-out], silk_cor$contribution[-out], method = "pearson")

```

## correlation between transcriptomics and proteomics - Hypoxia Medulla

```{r correlation medulla hypoxia}

# Subset transcriptomics data to protein detected
medu <- normalized_counts[["Medulla"]]
medu_df <- medu[which(medu$external_gene_name %in% hypoxia_cor$Gene),]

# Subset protein dataset to cortex
hypoxia_medulla <- hypoxia_cor[,c(1, 7:12)]

medulla <- merge(hypoxia_medulla, medu_df, by.x = "Gene", by.y = "external_gene_name")
medulla <- medulla[, -c(8,24:25)]

# Read the metadata and keep only the samples that match the proteomics
info <- read.csv("/Users/tianyili/Desktop/Ovarioid/data/processed/info_Medulla.csv", sep = ";")
info_med <- info %>% select(name, culture) %>% 
  filter(culture != "2D")
info_med[11:16,1] <- colnames(angio_medulla)[2:7]
info_med[11:16,2] <- c("Tissue", "Tissue", "Tissue", "3D", "3D", "3D")

# Remove samples not used for proteomics "P28012_1008_S8"  "P28012_1005_S5"  "P28012_1019_S19" "P28012_1016_S16"
dif <- c("P28012_1008_S8", "P28012_1005_S5", "P28012_1019_S19", "P28012_1016_S16")
info_med <- info_med[-which(info_med$name %in% dif),]
info_med <- info_med %>% arrange(culture)

rownames(medulla) <- medulla$Gene
medulla <- medulla[,which(colnames(medulla) %in% info_med$name)]
medulla$Gene <- rownames(medulla)

med_protein <- medulla[,c(1:6,13)] %>% 
  tidyr::pivot_longer(!Gene, names_to = "sample", values_to = "contribution")

med_RNA <- medulla[, 7:13]
id <- match(colnames(med_RNA), c("P28012_1030_S30", "P28012_1046_S46", "P28012_1061_S61",
                           "P28012_1035_S35", "P28012_1051_S51", "P28012_1066_S66", "Gene"))
med_RNA <- med_RNA[, id]

medulla_RNA <- med_RNA
medulla_RNA <- medulla_RNA %>% 
  tidyr::pivot_longer(!Gene, names_to = "sample", values_to = "contribution")
colnames(medulla_RNA) <- paste0(colnames(medulla_RNA), "_RNA")

medul <- cbind(medulla_RNA, med_protein)
medul <- medul %>% filter(contribution > 0)
medul$type <- ifelse(str_detect(medul$sample, "T"), "Tissue", "Silk-Ov")

p4 <- ggplot(medul, aes(x = contribution_RNA, y = contribution, col = type)) + 
  geom_point(aes(shape = type)) + theme_bw() +
  geom_smooth(method = "lm", se = F) + scale_color_manual(values = c("#1B7C3D", "#2B6A99")) +
  ggtitle("Hypoxia - Medulla correlation") + xlim(c(0, 100000))

# Correlation test
tis_med <- medul %>% filter(type == "Tissue")

# Remove outlier
out <- which(tis_med$contribution_RNA > 100000)
out # no outlier
cor.test(tis_med$contribution_RNA, tis_med$contribution, method = "pearson")

silk_med <- medul %>% filter(type != "Tissue")

# Remove outlier
out <- which(silk_med$contribution_RNA > 100000)
out # no outlier
cor.test(silk_med$contribution_RNA[-out], silk_med$contribution[-out], method = "pearson")


```

## Combined correlation fig

```{r combine fig}

p3 + p4 + p1 + p2

```

## Individual protein plotting

```{r individual protein}

# Remove outlier if not removed before
cor <- cor[,-ncol(cor)]

d1 <- cor %>% filter(Gene %in% c("COL1A1", "LAMA1")) 

d1 <- d1 %>% select(!WikiPathways.x) %>% 
  pivot_longer(!Gene, names_to = "Sample", values_to = "value")
d1$culture <- c(rep("Tissue", 3), rep("Silk-Ov", 2))
d1$culture <- factor(d1$culture, levels = c("Tissue", "Silk-Ov"))

ggplot(d1, aes(x = culture, y = value)) + geom_boxplot() + theme_bw() +
  geom_point() +
  ggtitle("COL1A1") + ylab("Normalized Protein Contribution") + xlab("")

## Medulla

d2 <- med %>% filter(Gene %in% c("COL1A1", "LAMA1")) 

d2 <- d2 %>% select(!WikiPathways.x) %>% 
  pivot_longer(!Gene, names_to = "Sample", values_to = "value")
d2$culture <- c(rep("Tissue", 3), rep("Silk-Ov", 3))
d2$culture <- factor(d2$culture, levels = c("Tissue", "Silk-Ov"))

ggplot(d2, aes(x = culture, y = value)) + geom_boxplot() + theme_bw() +
  ggtitle("COL1A1") + ylab("Normalized Protein Contribution") + xlab("")

```

## Session info

```{r session info}

sessionInfo()

```

